{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fd9aca63",
   "metadata": {},
   "source": [
    "# Creating a dataset of the positive abstract set"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02f2db87",
   "metadata": {},
   "source": [
    "Goals of the labeling redo:\n",
    "1. Read in the dataset\n",
    "2. Turn it into list of sentences\n",
    "3. Map a label to each word\n",
    "    - Incorporate Disease Lookup from GARD database\n",
    "    - Location is same\n",
    "    - Epidemiologic Identifier {Prevalence, Incidence, Occurrence}\n",
    "    - Do tokenization with just the main location spacy model so there is no mismatch\n",
    "    - Do independent statistic labeling (ranges but not if 95%CI)\n",
    "4. Randomly split the dataset into test (50 abstract) set, and then train/val sets with train-test split - incorporate seed\n",
    "5. Save into the correct format    \n",
    "6. Manually fix the test set (Separate)\n",
    "7. After training the transformer - do sentence level extraction if there is at least one relevant STAT tag"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d991b086",
   "metadata": {},
   "source": [
    "## (1) Reading in the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d54d5046",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting en-core-web-trf==3.1.0\n",
      "  Downloading https://github.com/explosion/spacy-models/releases/download/en_core_web_trf-3.1.0/en_core_web_trf-3.1.0-py3-none-any.whl (460.2 MB)\n",
      "\u001b[K     |████████████████████████████████| 460.2 MB 13 kB/s s eta 0:00:01             | 9.7 MB 6.6 MB/s eta 0:01:09        | 17.5 MB 6.6 MB/s eta 0:01:08:01:07[K     |█▊                              | 24.7 MB 6.6 MB/s eta 0:01:07                          | 28.7 MB 6.6 MB/s eta 0:01:0633.1 MB 6.6 MB/s eta 0:01:05| 36.8 MB 6.6 MB/s eta 0:01:05                        | 39.9 MB 6.6 MB/s eta 0:01:04 eta 0:01:04.2 MB 6.6 MB/s eta 0:01:01    |████▋                           | 66.2 MB 75.7 MB/s eta 0:00:06          | 73.7 MB 75.7 MB/s eta 0:00:06�███▊                          | 81.6 MB 75.7 MB/s eta 0:00:05    |██████                          | 84.7 MB 75.7 MB/s eta 0:00:0588.8 MB 75.7 MB/s eta 0:00:05:00:055██                         | 100.4 MB 75.7 MB/s eta 0:00:05�▎                        | 104.3 MB 75.7 MB/s eta 0:00:05███████▌                        | 107.7 MB 75.7 MB/s eta 0:00:055.7 MB/s eta 0:00:05████████                        | 115.6 MB 75.7 MB/s eta 0:00:05                    | 119.3 MB 75.7 MB/s eta 0:00:053.5 MB 75.7 MB/s eta 0:00:05�███▉                       | 127.2 MB 75.7 MB/s eta 0:00:05131.1 MB 75.7 MB/s eta 0:00:05          | 134.6 MB 75.7 MB/s eta 0:00:05�▋                      | 138.0 MB 67.7 MB/s eta 0:00:05MB/s eta 0:00:057 MB/s eta 0:00:05 eta 0:00:05█████████                     | 157.3 MB 67.7 MB/s eta 0:00:05        | 161.1 MB 67.7 MB/s eta 0:00:05| 164.2 MB 67.7 MB/s eta 0:00:05��                    | 167.9 MB 67.7 MB/s eta 0:00:057 MB/s eta 0:00:05        | 190.4 MB 67.7 MB/s eta 0:00:04        | 194.2 MB 67.7 MB/s eta 0:00:04��██                  | 201.7 MB 67.7 MB/s eta 0:00:04�█████████▎                 | 205.7 MB 67.7 MB/s eta 0:00:04�█▋                 | 209.4 MB 82.7 MB/s eta 0:00:04| 212.7 MB 82.7 MB/s eta 0:00:03�██                 | 217.2 MB 82.7 MB/s eta 0:00:03 |███████████████▍                | 220.8 MB 82.7 MB/s eta 0:00:034.8 MB 82.7 MB/s eta 0:00:03          | 228.6 MB 82.7 MB/s eta 0:00:03��▍               | 236.4 MB 82.7 MB/s eta 0:00:03�               | 239.9 MB 82.7 MB/s eta 0:00:03        | 243.7 MB 82.7 MB/s eta 0:00:03ta 0:00:03█████████████████▌              | 251.2 MB 82.7 MB/s eta 0:00:03�▊              | 255.3 MB 82.7 MB/s eta 0:00:03��██████████████              | 258.9 MB 82.7 MB/s eta 0:00:03| 262.8 MB 82.7 MB/s eta 0:00:03MB/s eta 0:00:03▉             | 270.2 MB 82.7 MB/s eta 0:00:03MB/s eta 0:00:03 82.7 MB/s eta 0:00:03| 281.8 MB 73.7 MB/s eta 0:00:03███████████████████▉            | 285.3 MB 73.7 MB/s eta 0:00:03��██            | 289.1 MB 73.7 MB/s eta 0:00:030:03██████████████▋           | 296.8 MB 73.7 MB/s eta 0:00:03███████████████           | 300.9 MB 73.7 MB/s eta 0:00:03:00:03          | 308.7 MB 73.7 MB/s eta 0:00:03:00:03:00:02████████████▎         | 320.4 MB 73.7 MB/s eta 0:00:02��███████████▎        | 335.0 MB 73.7 MB/s eta 0:00:02███████████████████▌        | 338.7 MB 73.7 MB/s eta 0:00:02�██████████▉        | 343.0 MB 73.7 MB/s eta 0:00:024 MB 73.7 MB/s eta 0:00:02�███████████▍       | 350.0 MB 77.1 MB/s eta 0:00:02�███████████▋       | 353.6 MB 77.1 MB/s eta 0:00:02�███████████▉       | 357.3 MB 77.1 MB/s eta 0:00:02��█████████▏      | 361.4 MB 77.1 MB/s eta 0:00:02MB/s eta 0:00:02�██▋      | 368.4 MB 77.1 MB/s eta 0:00:02�██████████████████▉      | 372.0 MB 77.1 MB/s eta 0:00:02██████████████████████████▏     | 375.9 MB 77.1 MB/s eta 0:00:02MB/s eta 0:00:02��██▋     | 383.4 MB 77.1 MB/s eta 0:00:01��███████████████████     | 386.8 MB 77.1 MB/s eta 0:00:01███████▏    | 390.8 MB 77.1 MB/s eta 0:00:01��█████████████▌    | 394.7 MB 77.1 MB/s eta 0:00:01ta 0:00:01██████████████████▏   | 405.4 MB 77.1 MB/s eta 0:00:01██████████████████▍   | 409.1 MB 77.1 MB/s eta 0:00:01��█████   | 415.6 MB 77.1 MB/s eta 0:00:01ta 0:00:01422.8 MB 80.1 MB/s eta 0:00:010.1 MB/s eta 0:00:01ta 0:00:01MB 80.1 MB/s eta 0:00:01�█▍ | 436.8 MB 80.1 MB/s eta 0:00:01�█▋ | 440.1 MB 80.1 MB/s eta 0:00:01�█▉ | 443.8 MB 80.1 MB/s eta 0:00:01�██ | 447.2 MB 80.1 MB/s eta 0:00:01��███████████▍| 450.8 MB 80.1 MB/s eta 0:00:01��███████████▋| 454.2 MB 80.1 MB/s eta 0:00:0101\n",
      "\u001b[?25hRequirement already satisfied: spacy<3.2.0,>=3.1.0 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from en-core-web-trf==3.1.0) (3.1.2)\n",
      "Collecting spacy-transformers<1.1.0,>=1.0.3\n",
      "  Downloading spacy_transformers-1.0.6-py2.py3-none-any.whl (42 kB)\n",
      "\u001b[K     |████████████████████████████████| 42 kB 858 kB/s  eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: setuptools in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (58.0.4)\n",
      "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (21.0)\n",
      "Requirement already satisfied: typer<0.4.0,>=0.3.0 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (0.3.2)\n",
      "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (2.0.5)\n",
      "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (1.0.5)\n",
      "Requirement already satisfied: pathy>=0.3.5 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (0.6.0)\n",
      "Requirement already satisfied: numpy>=1.15.0 in /usr/local/lib64/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (1.19.5)\n",
      "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (4.62.2)\n",
      "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.7 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (3.0.8)\n",
      "Requirement already satisfied: blis<0.8.0,>=0.4.0 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (0.7.4)\n",
      "Requirement already satisfied: srsly<3.0.0,>=2.4.1 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (2.4.1)\n",
      "Requirement already satisfied: catalogue<2.1.0,>=2.0.4 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (2.0.6)\n",
      "Requirement already satisfied: requests<3.0.0,>=2.13.0 in /usr/local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (2.26.0)\n",
      "Requirement already satisfied: pydantic!=1.8,!=1.8.1,<1.9.0,>=1.7.4 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (1.8.2)\n",
      "Requirement already satisfied: jinja2 in /usr/local/lib64/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (3.0.1)\n",
      "Requirement already satisfied: typing-extensions<4.0.0.0,>=3.7.4 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (3.7.4.3)\n",
      "Requirement already satisfied: thinc<8.1.0,>=8.0.8 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (8.0.8)\n",
      "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (3.0.5)\n",
      "Requirement already satisfied: wasabi<1.1.0,>=0.8.1 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (0.8.2)\n",
      "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.6/site-packages (from catalogue<2.1.0,>=2.0.4->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (3.5.0)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /usr/local/lib/python3.6/site-packages (from packaging>=20.0->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (2.4.7)\n",
      "Requirement already satisfied: dataclasses<1.0,>=0.6 in /usr/local/lib/python3.6/site-packages (from pathy>=0.3.5->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (0.8)\n",
      "Requirement already satisfied: smart-open<6.0.0,>=5.0.0 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from pathy>=0.3.5->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (5.2.1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.6/site-packages (from requests<3.0.0,>=2.13.0->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (3.2)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/site-packages (from requests<3.0.0,>=2.13.0->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (2021.5.30)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.6/site-packages (from requests<3.0.0,>=2.13.0->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (1.26.6)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /usr/local/lib/python3.6/site-packages (from requests<3.0.0,>=2.13.0->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (2.0.4)\n",
      "Requirement already satisfied: torch>=1.5.0 in /usr/local/lib64/python3.6/site-packages (from spacy-transformers<1.1.0,>=1.0.3->en-core-web-trf==3.1.0) (1.9.0)\n",
      "Collecting spacy-alignments<1.0.0,>=0.7.2\n",
      "  Downloading spacy_alignments-0.8.3-cp36-cp36m-manylinux2014_x86_64.whl (998 kB)\n",
      "\u001b[K     |████████████████████████████████| 998 kB 30.6 MB/s eta 0:00:01\n",
      "\u001b[?25hCollecting transformers<4.10.0,>=3.4.0\n",
      "  Downloading transformers-4.9.2-py3-none-any.whl (2.6 MB)\n",
      "\u001b[K     |████████████████████████████████| 2.6 MB 51.2 MB/s eta 0:00:01███████████████████████████     | 2.2 MB 51.2 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: contextvars<3,>=2.4 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from thinc<8.1.0,>=8.0.8->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (2.4)\n",
      "Requirement already satisfied: immutables>=0.9 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from contextvars<3,>=2.4->thinc<8.1.0,>=8.0.8->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (0.16)\n",
      "Requirement already satisfied: tokenizers<0.11,>=0.10.1 in /usr/local/lib64/python3.6/site-packages (from transformers<4.10.0,>=3.4.0->spacy-transformers<1.1.0,>=1.0.3->en-core-web-trf==3.1.0) (0.10.3)\n",
      "Collecting huggingface-hub==0.0.12\n",
      "  Downloading huggingface_hub-0.0.12-py3-none-any.whl (37 kB)\n",
      "Requirement already satisfied: filelock in /usr/local/lib/python3.6/site-packages (from transformers<4.10.0,>=3.4.0->spacy-transformers<1.1.0,>=1.0.3->en-core-web-trf==3.1.0) (3.0.12)\n",
      "Requirement already satisfied: sacremoses in /usr/local/lib/python3.6/site-packages (from transformers<4.10.0,>=3.4.0->spacy-transformers<1.1.0,>=1.0.3->en-core-web-trf==3.1.0) (0.0.45)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /usr/local/lib64/python3.6/site-packages (from transformers<4.10.0,>=3.4.0->spacy-transformers<1.1.0,>=1.0.3->en-core-web-trf==3.1.0) (5.4.1)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib64/python3.6/site-packages (from transformers<4.10.0,>=3.4.0->spacy-transformers<1.1.0,>=1.0.3->en-core-web-trf==3.1.0) (2021.8.28)\n",
      "Requirement already satisfied: importlib-metadata in /usr/local/lib/python3.6/site-packages (from transformers<4.10.0,>=3.4.0->spacy-transformers<1.1.0,>=1.0.3->en-core-web-trf==3.1.0) (4.6.4)\n",
      "Requirement already satisfied: click<7.2.0,>=7.1.1 in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from typer<0.4.0,>=0.3.0->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (7.1.2)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib64/python3.6/site-packages (from jinja2->spacy<3.2.0,>=3.1.0->en-core-web-trf==3.1.0) (2.0.1)\n",
      "Requirement already satisfied: joblib in /usr/local/lib/python3.6/site-packages (from sacremoses->transformers<4.10.0,>=3.4.0->spacy-transformers<1.1.0,>=1.0.3->en-core-web-trf==3.1.0) (1.0.1)\n",
      "Requirement already satisfied: six in /home/wzkariampuzha/.local/lib/python3.6/site-packages (from sacremoses->transformers<4.10.0,>=3.4.0->spacy-transformers<1.1.0,>=1.0.3->en-core-web-trf==3.1.0) (1.15.0)\n",
      "Installing collected packages: huggingface-hub, transformers, spacy-alignments, spacy-transformers, en-core-web-trf\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "datasets 1.12.1 requires huggingface-hub<0.1.0,>=0.0.14, but you have huggingface-hub 0.0.12 which is incompatible.\u001b[0m\n",
      "Successfully installed en-core-web-trf-3.1.0 huggingface-hub-0.0.12 spacy-alignments-0.8.3 spacy-transformers-1.0.6 transformers-4.9.2\n",
      "\u001b[33mWARNING: You are using pip version 21.2.4; however, version 21.3 is available.\n",
      "You should consider upgrading via the '/usr/bin/python3 -m pip install --upgrade pip' command.\u001b[0m\n",
      "\u001b[38;5;2m✔ Download and installation successful\u001b[0m\n",
      "You can now load the package via spacy.load('en_core_web_trf')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/wzkariampuzha/.local/lib/python3.6/site-packages/spacy/util.py:732: UserWarning: [W095] Model 'en_ner_bc5cdr_md' (0.4.0) was trained with spaCy v3.0 and may not be 100% compatible with the current version (3.1.2). If you see errors or degraded performance, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy validate\n",
      "  warnings.warn(warn_msg)\n"
     ]
    }
   ],
   "source": [
    "#Download all dependencies, only need to do this once\n",
    "#import sys\n",
    "#!{sys.executable} -m pip install spacy\n",
    "#!{sys.executable} -m spacy download en_core_web_lg\n",
    "#!{sys.executable} -m spacy download en_core_web_trf\n",
    "#!{sys.executable} -m pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.4.0/en_ner_bc5cdr_md-0.4.0.tar.gz\n",
    "#!{sys.executable} -m pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.4.0/en_ner_bionlp13cg_md-0.4.0.tar.gz\n",
    "\n",
    "import nltk\n",
    "#nltk.download('stopwords')\n",
    "#nltk.download('punkt')\n",
    "from nltk.corpus import stopwords\n",
    "from nltk import tokenize\n",
    "STOPWORDS = set(stopwords.words('english'))\n",
    "import string\n",
    "PUNCTUATION = set(char for char in string.punctuation)\n",
    "import csv\n",
    "import spacy\n",
    "import re\n",
    "import pandas as pd\n",
    "from termcolor import colored\n",
    "from spacy import displacy\n",
    "from nltk.tag import StanfordNERTagger\n",
    "st = StanfordNERTagger('stanford-ner/classifiers/english.all.3class.distsim.crf.ser.gz',\n",
    "                        'stanford-ner/stanford-ner.jar',\n",
    "                        encoding='utf-8')\n",
    "nlp = spacy.load('en_core_web_trf')\n",
    "nlpSci = spacy.load(\"en_ner_bc5cdr_md\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e81526af",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>pmid</th>\n",
       "      <th>abstract</th>\n",
       "      <th>epi_prob</th>\n",
       "      <th>is_epi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>750</th>\n",
       "      <td>CD4 deficiency</td>\n",
       "      <td>34128866</td>\n",
       "      <td>&lt;h4&gt;Abstract&lt;/h4&gt;Previous studies have suggest...</td>\n",
       "      <td>0.519289</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>751</th>\n",
       "      <td>Muscle eye brain disease</td>\n",
       "      <td>31344241</td>\n",
       "      <td>&lt;h4&gt;Background and objective&lt;/h4&gt;To understand...</td>\n",
       "      <td>0.514754</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>Enchondromatosis dwarfism deafness</td>\n",
       "      <td>32355641</td>\n",
       "      <td>Anticancer drug nephrotoxicity is an important...</td>\n",
       "      <td>0.512264</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753</th>\n",
       "      <td>Severe congenital neutropenia autosomal dominant</td>\n",
       "      <td>33554218</td>\n",
       "      <td>Coronavirus disease 2019 (COVID-19) is emergin...</td>\n",
       "      <td>0.511204</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754</th>\n",
       "      <td>Palindromic rheumatism</td>\n",
       "      <td>33780753</td>\n",
       "      <td>To date, retinal implants are the only availab...</td>\n",
       "      <td>0.500894</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              disease      pmid  \\\n",
       "750                                    CD4 deficiency  34128866   \n",
       "751                          Muscle eye brain disease  31344241   \n",
       "752                Enchondromatosis dwarfism deafness  32355641   \n",
       "753  Severe congenital neutropenia autosomal dominant  33554218   \n",
       "754                            Palindromic rheumatism  33780753   \n",
       "\n",
       "                                              abstract  epi_prob  is_epi  \n",
       "750  <h4>Abstract</h4>Previous studies have suggest...  0.519289    True  \n",
       "751  <h4>Background and objective</h4>To understand...  0.514754    True  \n",
       "752  Anticancer drug nephrotoxicity is an important...  0.512264    True  \n",
       "753  Coronavirus disease 2019 (COVID-19) is emergin...  0.511204    True  \n",
       "754  To date, retinal implants are the only availab...  0.500894    True  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"positive_abstract_set.csv\")\n",
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c3e46c3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "755 620\n"
     ]
    }
   ],
   "source": [
    "#Total pmids, unique pmids\n",
    "print(len(df[\"pmid\"]),len(df[\"pmid\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "85fc8e58",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "755 347\n"
     ]
    }
   ],
   "source": [
    "#total abstracts, total diseases, was supposed to be a sample of 500 diseases, but obviously only 347 have epi studies\n",
    "print(len(df[\"disease\"]),len(df[\"disease\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0a9f093c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>pmid</th>\n",
       "      <th>abstract</th>\n",
       "      <th>epi_prob</th>\n",
       "      <th>is_epi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Deafness, autosomal dominant nonsyndromic sens...</td>\n",
       "      <td>33962637</td>\n",
       "      <td>&lt;h4&gt;Background&lt;/h4&gt;The incidence of hydrocepha...</td>\n",
       "      <td>0.998725</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Anencephaly and spina bifida X-linked</td>\n",
       "      <td>33962637</td>\n",
       "      <td>&lt;h4&gt;Background&lt;/h4&gt;The incidence of hydrocepha...</td>\n",
       "      <td>0.998725</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Distal renal tubular acidosis with hemolytic a...</td>\n",
       "      <td>34167483</td>\n",
       "      <td>&lt;h4&gt;Background&lt;/h4&gt;The prevalence of Multiple ...</td>\n",
       "      <td>0.998105</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Cardiomyopathy due to anthracyclines</td>\n",
       "      <td>34167483</td>\n",
       "      <td>&lt;h4&gt;Background&lt;/h4&gt;The prevalence of Multiple ...</td>\n",
       "      <td>0.998105</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Syndactyly Cenani Lenz type</td>\n",
       "      <td>28402072</td>\n",
       "      <td>&lt;h4&gt;Background&lt;/h4&gt;Unilateral lung agenesis is...</td>\n",
       "      <td>0.998031</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>Cryoglobulinemic vasculitis</td>\n",
       "      <td>33526985</td>\n",
       "      <td>&lt;b&gt;Background:&lt;/b&gt; Osteogenesis imperfecta (OI...</td>\n",
       "      <td>0.556312</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>743</th>\n",
       "      <td>Infantile onset spinocerebellar ataxia</td>\n",
       "      <td>33602758</td>\n",
       "      <td>Q fever can present as a fever of unknown aeti...</td>\n",
       "      <td>0.556216</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>744</th>\n",
       "      <td>Progressive non-fluent aphasia</td>\n",
       "      <td>33602758</td>\n",
       "      <td>Q fever can present as a fever of unknown aeti...</td>\n",
       "      <td>0.556216</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>Nephropathy, deafness, and hyperparathyroidism</td>\n",
       "      <td>34128866</td>\n",
       "      <td>&lt;h4&gt;Abstract&lt;/h4&gt;Previous studies have suggest...</td>\n",
       "      <td>0.519289</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>750</th>\n",
       "      <td>CD4 deficiency</td>\n",
       "      <td>34128866</td>\n",
       "      <td>&lt;h4&gt;Abstract&lt;/h4&gt;Previous studies have suggest...</td>\n",
       "      <td>0.519289</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               disease      pmid  \\\n",
       "0    Deafness, autosomal dominant nonsyndromic sens...  33962637   \n",
       "1                Anencephaly and spina bifida X-linked  33962637   \n",
       "10   Distal renal tubular acidosis with hemolytic a...  34167483   \n",
       "11                Cardiomyopathy due to anthracyclines  34167483   \n",
       "15                         Syndactyly Cenani Lenz type  28402072   \n",
       "..                                                 ...       ...   \n",
       "742                        Cryoglobulinemic vasculitis  33526985   \n",
       "743             Infantile onset spinocerebellar ataxia  33602758   \n",
       "744                     Progressive non-fluent aphasia  33602758   \n",
       "749     Nephropathy, deafness, and hyperparathyroidism  34128866   \n",
       "750                                     CD4 deficiency  34128866   \n",
       "\n",
       "                                              abstract  epi_prob  is_epi  \n",
       "0    <h4>Background</h4>The incidence of hydrocepha...  0.998725    True  \n",
       "1    <h4>Background</h4>The incidence of hydrocepha...  0.998725    True  \n",
       "10   <h4>Background</h4>The prevalence of Multiple ...  0.998105    True  \n",
       "11   <h4>Background</h4>The prevalence of Multiple ...  0.998105    True  \n",
       "15   <h4>Background</h4>Unilateral lung agenesis is...  0.998031    True  \n",
       "..                                                 ...       ...     ...  \n",
       "742  <b>Background:</b> Osteogenesis imperfecta (OI...  0.556312    True  \n",
       "743  Q fever can present as a fever of unknown aeti...  0.556216    True  \n",
       "744  Q fever can present as a fever of unknown aeti...  0.556216    True  \n",
       "749  <h4>Abstract</h4>Previous studies have suggest...  0.519289    True  \n",
       "750  <h4>Abstract</h4>Previous studies have suggest...  0.519289    True  \n",
       "\n",
       "[246 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pmid = df[\"pmid\"]\n",
    "df[pmid.isin(pmid[pmid.duplicated()])]#.sort(\"pmid\")\n",
    "#This shows that 1) duplicate pmids are tagged with different diseases, but more importantly that 2) the disease that is \n",
    "#labeled was just from the search and may not have anything to do with the actual abstract (see https://pubmed.ncbi.nlm.nih.gov/33602758/ in row 743 & 744)\n",
    "#3) This means that we need to do disease extraction too and cannot rely on the search disease that is labeled.\n",
    "#4) It also means that duplicates can be dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c7656308",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>34128866</td>\n",
       "      <td>&lt;h4&gt;Abstract&lt;/h4&gt;Previous studies have suggest...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>31344241</td>\n",
       "      <td>&lt;h4&gt;Background and objective&lt;/h4&gt;To understand...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>32355641</td>\n",
       "      <td>Anticancer drug nephrotoxicity is an important...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>33554218</td>\n",
       "      <td>Coronavirus disease 2019 (COVID-19) is emergin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>619</th>\n",
       "      <td>33780753</td>\n",
       "      <td>To date, retinal implants are the only availab...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         pmid                                           abstract\n",
       "615  34128866  <h4>Abstract</h4>Previous studies have suggest...\n",
       "616  31344241  <h4>Background and objective</h4>To understand...\n",
       "617  32355641  Anticancer drug nephrotoxicity is an important...\n",
       "618  33554218  Coronavirus disease 2019 (COVID-19) is emergin...\n",
       "619  33780753  To date, retinal implants are the only availab..."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Gather only necessary information\n",
    "abstracts_df = df.drop_duplicates(subset='pmid', keep='first', ignore_index=True)\n",
    "abstracts_df = abstracts_df[['pmid','abstract']]\n",
    "abstracts_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a173d885",
   "metadata": {},
   "source": [
    "### HTML Removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3cd1c050",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_html(string):\n",
    "    #string = re.sub(':', '', string)\n",
    "    string = re.sub('<.{1,4}>', ' ', string)\n",
    "    #string = re.sub('.{1,3}>', ' ', string)\n",
    "    #string = re.sub('<.{1,3}', ' ', string)\n",
    "    string = re.sub(\"  *\", \" \" , string)\n",
    "    string = re.sub(\"^ \", \"\" , string)\n",
    "    string = re.sub(\" $\", \"\" , string)\n",
    "    string = re.sub(\"  \", \" \" , string)\n",
    "    #This also removes extra parentheses\n",
    "    #string= string.replace(\"(\", ' ')\n",
    "    #string= string.replace(\")\", ' ')\n",
    "    \n",
    "    string=string.strip()\n",
    "    return string"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c792439",
   "metadata": {},
   "source": [
    "### Jennifer's functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b916976d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def printHighlighted(doc, indices):\n",
    "    final = ''\n",
    "    start = 0\n",
    "    for i in indices:\n",
    "        final += doc[start:i[0]].text+' '\n",
    "        final += colored(doc[i[0]:i[1]].text, 'red', 'on_yellow', attrs=['bold']) + ' '\n",
    "        start = i[1]\n",
    "    final += doc[start:].text\n",
    "    print(final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "61bfccbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def removeDuplicates(a):\n",
    "    for i in range(len(a)-1,0,-1):\n",
    "        if a[i] == a[i-1]:\n",
    "            del a[i]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9697028c",
   "metadata": {},
   "source": [
    "#### Location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "98b9ab10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getLocsNltk(text):\n",
    "    tokenized_text = word_tokenize(text)\n",
    "    classified_text = st.tag(tokenized_text)\n",
    "    locs = set()\n",
    "\n",
    "    for word in classified_text:\n",
    "        if word[1] == 'LOCATION':\n",
    "            if word[0] not in locs:\n",
    "                locs.add(word[0])\n",
    "    \n",
    "    return locs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ef52e853",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getLocsSpacy(doc):\n",
    "    locs = {}\n",
    "    for ent in doc.ents:\n",
    "        if ent.label_ == 'GPE':\n",
    "            tokens = {token.text for token in ent}\n",
    "            if ent.text not in locs:\n",
    "                locs[ent.text] = tokens\n",
    "            else:\n",
    "                for t in tokens:\n",
    "                    if t not in locs[ent.text]:\n",
    "                        locs[ent.text].add(t)     \n",
    "    return locs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bd97b0ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getLocs(text):\n",
    "    doc = nlp(text)\n",
    "    \n",
    "    spacyLocs = getLocsSpacy(doc)\n",
    "    nltkLocs = getLocsNltk(text)\n",
    "    locs = []\n",
    "    \n",
    "    for entity in spacyLocs:\n",
    "        if len(spacyLocs[entity] & nltkLocs) != 0:\n",
    "            locs.append(entity)\n",
    "            \n",
    "    return locs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8a4f26a",
   "metadata": {},
   "source": [
    "#### STATS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "14c95e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getTokenChunkDict(doc):\n",
    "    chunks = [chunk for chunk in doc.noun_chunks]\n",
    "    tokenToChunk = {}\n",
    "    for chunk in chunks:\n",
    "        for i in range(chunk.start, chunk.end):\n",
    "            tokenToChunk[i] = [chunk.start, chunk.end]\n",
    "    return tokenToChunk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "80bd5099",
   "metadata": {},
   "outputs": [],
   "source": [
    "def isValidStat(token):\n",
    "    ancestors = {a.text.lower() for a in token.ancestors}\n",
    "    if 'ci' in ancestors or 'confidence' in ancestors or 'interval' in ancestors or 'p' in ancestors or 'p-value' in ancestors or 'type' in ancestors:\n",
    "        return False\n",
    "    if 'times' in ancestors:\n",
    "        return False\n",
    "    if token.text.lower() == 'one' and len(token.doc) > token.i + 1 and token.doc[token.i + 1].text == 'of':\n",
    "        return False\n",
    "    if token.ent_type_ == 'DATE':\n",
    "        return False\n",
    "    if token.ent_type_ in {'CARDINAL','QUANTITY'}:\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5ba84b86",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getStats(abst, display=False):\n",
    "    doc = nlp(abst)\n",
    "    indices = []\n",
    "    tokenToChunk = getTokenChunkDict(doc)\n",
    "    key_val_dz = []\n",
    "    \n",
    "    for sent in doc.sents:\n",
    "        keywords = []\n",
    "        values = []\n",
    "        dzs = []\n",
    "        \n",
    "        keywords_text = []\n",
    "        values_text = []\n",
    "        dzs_text = []\n",
    "        \n",
    "        sciSent = nlpSci(sent.text)\n",
    "        \n",
    "        for token in sent:\n",
    "            sciToken = nlpSci(token.text)[0]\n",
    "            if token.text.lower() in {'prevalence','incidence','frequency','PR','prevalences','occurrence'}:\n",
    "                if token.i in tokenToChunk:\n",
    "                    keywords.append(tokenToChunk[token.i])\n",
    "                else:\n",
    "                    keywords.append([token.i, token.i+1])\n",
    "            if isValidStat(token) or isValidStat(nlp(token.text)[0]):\n",
    "                if token.i in tokenToChunk:\n",
    "                    values.append(tokenToChunk[token.i])\n",
    "                else:\n",
    "                    values.append([token.i, token.i+1])\n",
    "        if keywords != [] and values != []:\n",
    "            for token in sciSent:\n",
    "                if token.ent_type_ == 'DISEASE':\n",
    "                    for token_reg in sent:\n",
    "                        if token_reg.text == token.text:\n",
    "                            if token_reg.i in tokenToChunk:\n",
    "                                dzs.append(tokenToChunk[token_reg.i])\n",
    "                            else:\n",
    "                                dzs.append([token_reg.i, token_reg.i+1])\n",
    "            \n",
    "            removeDuplicates(keywords)\n",
    "            removeDuplicates(values)\n",
    "            removeDuplicates(dzs)\n",
    "            for i in keywords:\n",
    "                keywords_text.append(doc[i[0]:i[1]])\n",
    "            for i in values:\n",
    "                values_text.append(doc[i[0]:i[1]])\n",
    "            for i in dzs:\n",
    "                dzs_text.append(doc[i[0]:i[1]])\n",
    "            key_val_dz.append((keywords_text, values_text, dzs_text))\n",
    "            indices += keywords\n",
    "            indices += values\n",
    "            indices += dzs\n",
    "    indices = sorted(indices)\n",
    "    removeDuplicates(indices)\n",
    "    if display:\n",
    "        printHighlighted(doc, indices)\n",
    "    return key_val_dz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b57941e1",
   "metadata": {},
   "source": [
    "### Testing cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "85d97780",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \u001b[1m\u001b[43m\u001b[31mIncidence\u001b[0m of \u001b[1m\u001b[43m\u001b[31mHansen's disease\u001b[0m in Olmsted County, Minnesota, was \u001b[1m\u001b[43m\u001b[31m2.6/million/year\u001b[0m .\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[([Incidence], [2.6/million/year], [Hansen's disease])]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sent = \"Incidence of Hansen's disease in Olmsted County, Minnesota, was 2.6/million/year.\"\n",
    "#doc = nlp(sent)\n",
    "getStats(sent, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9c8b6c69",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<h4>Background</h4>Most epidemiological data on vitiligo refer to selected environments or focus on the prevalence of comorbidity unrelated to the population.<h4>Objective</h4>Aim of the study was to gain robust representative prevalence data on vitiligo and on associated dermatologic comorbidity in the German adult population.<h4>Methods</h4>A dual population-based approach was applied with 1) primary data obtained between 2004 and 2014 from dermatological exams in the general working population; 2) claims data from a large German statutory health insurance, reference year 2010.<h4>Results</h4>In the working cohort (N = 121,783; 57% male; mean age 43 years), the prevalence of vitiligo was 0.77% (0.84% in men; 0.67% in women). In the claims data (N = 1,619,678; 38% male; mean age 46 years), prevalence was 0.17% (0.14% in men; 0.18% in women). In the working cohort, vitiligo was significantly more common in people with fair skin type, ephelides and port-wine stains and less common in people with acne and solar lentigines. In the claims data, vitiligo was associated with a variety of skin conditions, eg, atopic dermatitis, psoriasis and alopecia areata.<h4>Conclusion</h4>The resulting discrepancy of claims vs primary data between 0.17% and 0.77% indicates the most probable spectrum of vitiligo prevalence in Germany. It is more frequently observed in clinical exams than recorded in claims data, indicating a marked proportion of people seeking no medical help. Such nonattendance may result from the fact that many treatment options do not provide satisfying benefits to the patients.\n",
      "...............\n",
      "Background Most epidemiological data on vitiligo refer to selected environments or focus on the prevalence of comorbidity unrelated to the population. Objective Aim of the study was to gain robust representative prevalence data on vitiligo and on associated dermatologic comorbidity in the German adult population. Methods A dual population-based approach was applied with 1) primary data obtained between 2004 and 2014 from dermatological exams in the general working population; 2) claims data from a large German statutory health insurance, reference year 2010. Results In the working cohort (N = \u001b[1m\u001b[43m\u001b[31m121,783\u001b[0m ; \u001b[1m\u001b[43m\u001b[31m57\u001b[0m % male; mean age \u001b[1m\u001b[43m\u001b[31m43 years\u001b[0m ), \u001b[1m\u001b[43m\u001b[31mthe prevalence\u001b[0m of \u001b[1m\u001b[43m\u001b[31mvitiligo\u001b[0m was \u001b[1m\u001b[43m\u001b[31m0.77%\u001b[0m ( \u001b[1m\u001b[43m\u001b[31m0.84%\u001b[0m in men; \u001b[1m\u001b[43m\u001b[31m0.67%\u001b[0m in women). In the claims data (N = \u001b[1m\u001b[43m\u001b[31m1,619,678\u001b[0m ; \u001b[1m\u001b[43m\u001b[31m38% male\u001b[0m ; mean age \u001b[1m\u001b[43m\u001b[31m46\u001b[0m years), \u001b[1m\u001b[43m\u001b[31mprevalence\u001b[0m was \u001b[1m\u001b[43m\u001b[31m0.17%\u001b[0m ( \u001b[1m\u001b[43m\u001b[31m0.14%\u001b[0m in men; \u001b[1m\u001b[43m\u001b[31m0.18%\u001b[0m in women). In the working cohort, vitiligo was significantly more common in people with fair skin type, ephelides and port-wine stains and less common in people with acne and solar lentigines. In the claims data, vitiligo was associated with a variety of skin conditions, eg, atopic dermatitis, psoriasis and alopecia areata. Conclusion The resulting discrepancy of claims vs primary data between \u001b[1m\u001b[43m\u001b[31m0.17%\u001b[0m and \u001b[1m\u001b[43m\u001b[31m0.77%\u001b[0m indicates the most probable spectrum of \u001b[1m\u001b[43m\u001b[31mvitiligo prevalence\u001b[0m in Germany. It is more frequently observed in clinical exams than recorded in claims data, indicating a marked proportion of people seeking no medical help. Such nonattendance may result from the fact that many treatment options do not provide satisfying benefits to the patients.\n",
      "[([the prevalence], [121,783, 57, 43 years, 0.77%, 0.84%, 0.67%], [vitiligo]), ([prevalence], [1,619,678, 38% male, 46, 0.17%, 0.14%, 0.18%], []), ([vitiligo prevalence], [0.17%, 0.77%], [vitiligo prevalence])]\n",
      "\n",
      "<h4>Abstract</h4>Previous studies have suggested. Neuroendocrine tumors (NETs) rare neoplasms, estimated annual incidence 6.9/100 000. They arise cells diffuse endocrine system, mainly dispersed throughout gastrointestinal (GI), pancreatic, respiratory tracts. The incidence GI-NETs recently begun show steady increase. According Surveillance, Epidemiology, End Results database, 53% patients NETs present localized disease, 20% locoregional disease, 27% distant metastases time diagnosis. Surgery mainstay treatment locoregional GI-NETs. Endoscopic resection option well-differentiated early GI-NETs, thought rarely metastasize lymph nodes. A lesion technically difficult resect via endoscopy indication local resection (partial resection without lymph node dissection). GI-NETs possible lymph node metastasis indication enterectomy lymph node dissection. For NETs metastatic lesions, cytoreduction surgery control hormonal hypersecretion alleviate symptoms; therefore, cytoreduction surgery recommended. The indications surgery vary based organ NET arose; therefore, understanding patient's clinical state individualized treatment based characteristics patient's GI-NET needed. This review summarizes surgical treatments GI-NETs organ.\n",
      "...............\n",
      "Abstract Previous studies have suggested. \u001b[1m\u001b[43m\u001b[31mNeuroendocrine tumors\u001b[0m (NETs) \u001b[1m\u001b[43m\u001b[31mrare neoplasms\u001b[0m , estimated \u001b[1m\u001b[43m\u001b[31mannual incidence\u001b[0m  \u001b[1m\u001b[43m\u001b[31m6.9/100\u001b[0m  \u001b[1m\u001b[43m\u001b[31m000\u001b[0m . They arise cells diffuse endocrine system, mainly dispersed throughout gastrointestinal (GI), pancreatic, respiratory tracts. The incidence GI-NETs recently begun show steady increase. According Surveillance, Epidemiology, End Results database, 53% patients NETs present localized disease, 20% locoregional disease, 27% distant metastases time diagnosis. Surgery mainstay treatment locoregional GI-NETs. Endoscopic resection option well-differentiated early GI-NETs, thought rarely metastasize lymph nodes. A lesion technically difficult resect via endoscopy indication local resection (partial resection without lymph node dissection). GI-NETs possible lymph node metastasis indication enterectomy lymph node dissection. For NETs metastatic lesions, cytoreduction surgery control hormonal hypersecretion alleviate symptoms; therefore, cytoreduction surgery recommended. The indications surgery vary based organ NET arose; therefore, understanding patient's clinical state individualized treatment based characteristics patient's GI-NET needed. This review summarizes surgical treatments GI-NETs organ.\n",
      "[([annual incidence], [6.9/100, 000], [Neuroendocrine tumors, rare neoplasms])]\n",
      "\n",
      "<h4>Abstract</h4>Previous studies have suggested.  Aim: There are more than 50 inherited lysosomal storage diseases (LSDs), and this study examined the incidence of clinically diagnosed LSDs in Sweden. Methods: The number of patients diagnosed during 1980-2009 was compiled from the registries of the two Swedish diagnostic laboratories that cover the whole country. Results: We identified 433 patients during the 30-year period, with a total incidence of one in every 6100 births and identified fairly constant annual diagnoses during the last 20 years. Krabbe disease was the most common (one in 39 000) followed by Gaucher disease (one in 47 000), metachromatic leukodystrophy and Salla disease. Gaucher disease was more frequent in Sweden than other European countries, due to a founder effect of the mutation (p.L444P) in northern Sweden. Metachromatic leukodystrophy was one of the. Neuroendocrine tumors (NETs) are rare neoplasms, with an estimated annual incidence of 6.9/100 000. They arise from cells of the diffuse endocrine system\n",
      "...............\n",
      "Abstract Previous studies have suggested. Aim: There are \u001b[1m\u001b[43m\u001b[31mmore than 50 inherited lysosomal storage diseases\u001b[0m ( \u001b[1m\u001b[43m\u001b[31mLSDs\u001b[0m ), and this study examined \u001b[1m\u001b[43m\u001b[31mthe incidence\u001b[0m of \u001b[1m\u001b[43m\u001b[31mclinically diagnosed LSDs\u001b[0m in Sweden. Methods: The number of patients diagnosed during 1980-2009 was compiled from the registries of the two Swedish diagnostic laboratories that cover the whole country. Results: We identified \u001b[1m\u001b[43m\u001b[31m433 patients\u001b[0m during \u001b[1m\u001b[43m\u001b[31mthe 30-year period\u001b[0m , with \u001b[1m\u001b[43m\u001b[31ma total incidence\u001b[0m of \u001b[1m\u001b[43m\u001b[31mone\u001b[0m in \u001b[1m\u001b[43m\u001b[31mevery 6100 births\u001b[0m and identified fairly constant annual diagnoses during \u001b[1m\u001b[43m\u001b[31mthe last 20 years\u001b[0m . Krabbe disease was the most common (one in 39 000) followed by Gaucher disease (one in 47 000), metachromatic leukodystrophy and Salla disease. Gaucher disease was more frequent in Sweden than other European countries, due to a founder effect of the mutation (p.L444P) in northern Sweden. Metachromatic leukodystrophy was one of the. \u001b[1m\u001b[43m\u001b[31mNeuroendocrine tumors\u001b[0m (NETs) are \u001b[1m\u001b[43m\u001b[31mrare neoplasms\u001b[0m , with \u001b[1m\u001b[43m\u001b[31man estimated annual incidence\u001b[0m of \u001b[1m\u001b[43m\u001b[31m6.9/100\u001b[0m  \u001b[1m\u001b[43m\u001b[31m000\u001b[0m . They arise from cells of the diffuse endocrine system\n",
      "[([the incidence], [more than 50 inherited lysosomal storage diseases], [LSDs, clinically diagnosed LSDs]), ([a total incidence], [433 patients, the 30-year period, one, every 6100 births, the last 20 years], []), ([an estimated annual incidence], [6.9/100, 000], [Neuroendocrine tumors, rare neoplasms])]\n",
      "\n",
      "Frontal fibrosing alopecia (FFA) is a variant of lichen planopilaris (LPP) with characteristic band-like frontotemporal hairline involvement and eyebrow loss. It most commonly occurs in post-menopausal White women.<sup>1</sup> In skin of color (SOC) individuals, FFA is often misdiagnosed as traction alopecia (TA),<sup>2</sup> and little data exists regarding the presentation of FFA in the SOC patient population.<sup>3</sup> As FFA incidence continues to increase,<sup>4</sup> we aim to understand differences in the presentation of FFA between White and Black women in order to aid in the accurate and timely diagnosis as well as help inform prognosis and management.\n",
      "...............\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Frontal fibrosing alopecia (FFA) is a variant of lichen planopilaris (LPP) with characteristic band-like frontotemporal hairline involvement and eyebrow loss. It most commonly occurs in post-menopausal White women. 1 In skin of color (SOC) individuals, FFA is often misdiagnosed as traction alopecia (TA), 2 and little data exists regarding the presentation of FFA in the SOC patient population. \u001b[1m\u001b[43m\u001b[31m3\u001b[0m As \u001b[1m\u001b[43m\u001b[31mFFA incidence\u001b[0m continues to increase, \u001b[1m\u001b[43m\u001b[31m4\u001b[0m we aim to understand differences in the presentation of FFA between White and Black women in order to aid in the accurate and timely diagnosis as well as help inform prognosis and management.\n",
      "[([FFA incidence], [3, 4], [])]\n",
      "\n",
      "'Although benzothiazole and its derivatives (BTHs) are considered emerging contaminants in diverse environments and organisms, little information is available about their contamination profiles and health impact in ambient particles. In this study, an optimized method of ultrasound-assisted extraction coupled with the selected reaction monitoring (SRM) mode of GC-EI-MS/MS was applied to characterize and analyze PM<sub>2.5</sub>-bound BTHs from three cities of China (Guangzhou, Shanghai, and Taiyuan) during the winter of 2018. The total BTH concentration (ΣBTHs) in PM<sub>2.5</sub> samples from the three cities decreased in the order of Guangzhou > Shanghai > Taiyuan, independently of the PM<sub>2.5</sub> concentration. Despite the large variation in concentration of ΣBTHs in PM<sub>2.5</sub>, 2-hydroxybenzothiazole (OTH) was always the predominant compound among the PM<sub>2.5</sub>-bound BTHs and accounted for 50-80% of total BTHs in the three regions. Results from human exposure assessment and toxicity screening indicated that the outdoor exposure risk of PM<sub>2.5</sub>-bound BTHs in toddlers was much higher than in adults, especially for OTH. The developmental and reproduction toxicity of OTH was further explored in vivo and in vitro. Exposure of mouse embryonic stem cells (mESCs) to OTH for 48 h significantly increased the intracellular reactive oxygen species (ROS) and induced DNA damage and apoptosis via the functionally activating p53 expression. In addition, the growth and development of zebrafish embryos were found to be severely affected after OTH treatment. An overall metabolomics study was conducted on the exposed zebrafish larvae. The results indicated that exposure to OTH inhibited the phenylalanine hydroxylation reaction, which further increased the accumulation of toxic phenylpyruvate and acetylphenylalanine in zebrafish. These findings provide important insights into the contamination profiles of PM<sub>2.5</sub>-bound BTHs and emphasize the health risk of OTH.'\n",
      "...............\n",
      "'Although benzothiazole and its derivatives (BTHs) are considered emerging contaminants in diverse environments and organisms, little information is available about their contamination profiles and health impact in ambient particles. In this study, an optimized method of ultrasound-assisted extraction coupled with the selected reaction monitoring (SRM) mode of GC-EI-MS/MS was applied to characterize and analyze PM 2.5 -bound BTHs from three cities of China (Guangzhou, Shanghai, and Taiyuan) during the winter of 2018. The total BTH concentration (ΣBTHs) in PM 2.5 samples from the three cities decreased in the order of Guangzhou > Shanghai > Taiyuan, independently of the PM 2.5 concentration. Despite the large variation in concentration of ΣBTHs in PM 2.5 , 2-hydroxybenzothiazole (OTH) was always the predominant compound among the PM 2.5 -bound BTHs and accounted for 50-80% of total BTHs in the three regions. Results from human exposure assessment and toxicity screening indicated that the outdoor exposure risk of PM 2.5 -bound BTHs in toddlers was much higher than in adults, especially for OTH. The developmental and reproduction toxicity of OTH was further explored in vivo and in vitro. Exposure of mouse embryonic stem cells (mESCs) to OTH for 48 h significantly increased the intracellular reactive oxygen species (ROS) and induced DNA damage and apoptosis via the functionally activating p53 expression. In addition, the growth and development of zebrafish embryos were found to be severely affected after OTH treatment. An overall metabolomics study was conducted on the exposed zebrafish larvae. The results indicated that exposure to OTH inhibited the phenylalanine hydroxylation reaction, which further increased the accumulation of toxic phenylpyruvate and acetylphenylalanine in zebrafish. These findings provide important insights into the contamination profiles of PM 2.5 -bound BTHs and emphasize the health risk of OTH.'\n",
      "[]\n",
      "\n",
      "Incidence of the disease in Olmsted County, Minnesota, was 2.6/million/year.\n",
      "...............\n",
      " \u001b[1m\u001b[43m\u001b[31mIncidence\u001b[0m of the disease in Olmsted County, Minnesota, was \u001b[1m\u001b[43m\u001b[31m2.6/million/year\u001b[0m .\n",
      "[([Incidence], [2.6/million/year], [])]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "test_dict = {'string0':\"<h4>Background</h4>Most epidemiological data on vitiligo refer to selected environments or focus on the prevalence of comorbidity unrelated to the population.<h4>Objective</h4>Aim of the study was to gain robust representative prevalence data on vitiligo and on associated dermatologic comorbidity in the German adult population.<h4>Methods</h4>A dual population-based approach was applied with 1) primary data obtained between 2004 and 2014 from dermatological exams in the general working population; 2) claims data from a large German statutory health insurance, reference year 2010.<h4>Results</h4>In the working cohort (N = 121,783; 57% male; mean age 43 years), the prevalence of vitiligo was 0.77% (0.84% in men; 0.67% in women). In the claims data (N = 1,619,678; 38% male; mean age 46 years), prevalence was 0.17% (0.14% in men; 0.18% in women). In the working cohort, vitiligo was significantly more common in people with fair skin type, ephelides and port-wine stains and less common in people with acne and solar lentigines. In the claims data, vitiligo was associated with a variety of skin conditions, eg, atopic dermatitis, psoriasis and alopecia areata.<h4>Conclusion</h4>The resulting discrepancy of claims vs primary data between 0.17% and 0.77% indicates the most probable spectrum of vitiligo prevalence in Germany. It is more frequently observed in clinical exams than recorded in claims data, indicating a marked proportion of people seeking no medical help. Such nonattendance may result from the fact that many treatment options do not provide satisfying benefits to the patients.\",\n",
    "'string2':\"<h4>Abstract</h4>Previous studies have suggested.  Aim: There are more than 50 inherited lysosomal storage diseases (LSDs), and this study examined the incidence of clinically diagnosed LSDs in Sweden. Methods: The number of patients diagnosed during 1980-2009 was compiled from the registries of the two Swedish diagnostic laboratories that cover the whole country. Results: We identified 433 patients during the 30-year period, with a total incidence of one in every 6100 births and identified fairly constant annual diagnoses during the last 20 years. Krabbe disease was the most common (one in 39 000) followed by Gaucher disease (one in 47 000), metachromatic leukodystrophy and Salla disease. Gaucher disease was more frequent in Sweden than other European countries, due to a founder effect of the mutation (p.L444P) in northern Sweden. Metachromatic leukodystrophy was one of the. Neuroendocrine tumors (NETs) are rare neoplasms, with an estimated annual incidence of 6.9/100 000. They arise from cells of the diffuse endocrine system\",\n",
    "'string1':\"<h4>Abstract</h4>Previous studies have suggested. Neuroendocrine tumors (NETs) rare neoplasms, estimated annual incidence 6.9/100 000. They arise cells diffuse endocrine system, mainly dispersed throughout gastrointestinal (GI), pancreatic, respiratory tracts. The incidence GI-NETs recently begun show steady increase. According Surveillance, Epidemiology, End Results database, 53% patients NETs present localized disease, 20% locoregional disease, 27% distant metastases time diagnosis. Surgery mainstay treatment locoregional GI-NETs. Endoscopic resection option well-differentiated early GI-NETs, thought rarely metastasize lymph nodes. A lesion technically difficult resect via endoscopy indication local resection (partial resection without lymph node dissection). GI-NETs possible lymph node metastasis indication enterectomy lymph node dissection. For NETs metastatic lesions, cytoreduction surgery control hormonal hypersecretion alleviate symptoms; therefore, cytoreduction surgery recommended. The indications surgery vary based organ NET arose; therefore, understanding patient's clinical state individualized treatment based characteristics patient's GI-NET needed. This review summarizes surgical treatments GI-NETs organ.\",\n",
    "'string3':\"Frontal fibrosing alopecia (FFA) is a variant of lichen planopilaris (LPP) with characteristic band-like frontotemporal hairline involvement and eyebrow loss. It most commonly occurs in post-menopausal White women.<sup>1</sup> In skin of color (SOC) individuals, FFA is often misdiagnosed as traction alopecia (TA),<sup>2</sup> and little data exists regarding the presentation of FFA in the SOC patient population.<sup>3</sup> As FFA incidence continues to increase,<sup>4</sup> we aim to understand differences in the presentation of FFA between White and Black women in order to aid in the accurate and timely diagnosis as well as help inform prognosis and management.\",\n",
    "'string4':\"'Although benzothiazole and its derivatives (BTHs) are considered emerging contaminants in diverse environments and organisms, little information is available about their contamination profiles and health impact in ambient particles. In this study, an optimized method of ultrasound-assisted extraction coupled with the selected reaction monitoring (SRM) mode of GC-EI-MS/MS was applied to characterize and analyze PM<sub>2.5</sub>-bound BTHs from three cities of China (Guangzhou, Shanghai, and Taiyuan) during the winter of 2018. The total BTH concentration (ΣBTHs) in PM<sub>2.5</sub> samples from the three cities decreased in the order of Guangzhou > Shanghai > Taiyuan, independently of the PM<sub>2.5</sub> concentration. Despite the large variation in concentration of ΣBTHs in PM<sub>2.5</sub>, 2-hydroxybenzothiazole (OTH) was always the predominant compound among the PM<sub>2.5</sub>-bound BTHs and accounted for 50-80% of total BTHs in the three regions. Results from human exposure assessment and toxicity screening indicated that the outdoor exposure risk of PM<sub>2.5</sub>-bound BTHs in toddlers was much higher than in adults, especially for OTH. The developmental and reproduction toxicity of OTH was further explored in vivo and in vitro. Exposure of mouse embryonic stem cells (mESCs) to OTH for 48\\xa0h significantly increased the intracellular reactive oxygen species (ROS) and induced DNA damage and apoptosis via the functionally activating p53 expression. In addition, the growth and development of zebrafish embryos were found to be severely affected after OTH treatment. An overall metabolomics study was conducted on the exposed zebrafish larvae. The results indicated that exposure to OTH inhibited the phenylalanine hydroxylation reaction, which further increased the accumulation of toxic phenylpyruvate and acetylphenylalanine in zebrafish. These findings provide important insights into the contamination profiles of PM<sub>2.5</sub>-bound BTHs and emphasize the health risk of OTH.'\",\n",
    "'string5':'Incidence of the disease in Olmsted County, Minnesota, was 2.6/million/year.'\n",
    "            }\n",
    "for i in range(6):\n",
    "    print(test_dict['string'+str(i)])\n",
    "    print('...............')\n",
    "    print(getStats(remove_html(test_dict['string'+str(i)]),True))\n",
    "    print('')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d92854b",
   "metadata": {},
   "source": [
    "## (3) Map label onto each word (done rule-based at the sentence level)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fca63c7",
   "metadata": {},
   "source": [
    "### Manual GARD Disease Look-up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0616419a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d.gard_id</th>\n",
       "      <th>d.name</th>\n",
       "      <th>d.synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6056</th>\n",
       "      <td>GARD:0013731</td>\n",
       "      <td>T-cell prolymphocytic leukemia</td>\n",
       "      <td>[T Cell Prolymphocytic Leukemia]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6057</th>\n",
       "      <td>GARD:0013735</td>\n",
       "      <td>Spastic paraplegia 47</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6058</th>\n",
       "      <td>GARD:0013737</td>\n",
       "      <td>AP-4-Associated Hereditary Spastic Paraplegia</td>\n",
       "      <td>[Severe intellectual disability and progressiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6059</th>\n",
       "      <td>GARD:0013743</td>\n",
       "      <td>Multicentric osteolysis, nodulosis and arthrop...</td>\n",
       "      <td>[Torg-Winchester Syndrome,Torg Syndrome,Nodulo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6060</th>\n",
       "      <td>GARD:0013818</td>\n",
       "      <td>Sphingosine phosphate lyase insufficiency synd...</td>\n",
       "      <td>[SPL insufficiency syndrome,SPLIS,Familial ste...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         d.gard_id                                             d.name  \\\n",
       "6056  GARD:0013731                     T-cell prolymphocytic leukemia   \n",
       "6057  GARD:0013735                              Spastic paraplegia 47   \n",
       "6058  GARD:0013737      AP-4-Associated Hereditary Spastic Paraplegia   \n",
       "6059  GARD:0013743  Multicentric osteolysis, nodulosis and arthrop...   \n",
       "6060  GARD:0013818  Sphingosine phosphate lyase insufficiency synd...   \n",
       "\n",
       "                                             d.synonyms  \n",
       "6056                   [T Cell Prolymphocytic Leukemia]  \n",
       "6057                                                NaN  \n",
       "6058  [Severe intellectual disability and progressiv...  \n",
       "6059  [Torg-Winchester Syndrome,Torg Syndrome,Nodulo...  \n",
       "6060  [SPL insufficiency syndrome,SPLIS,Familial ste...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Read in GARD diseases\n",
    "GARD_df = pd.read_csv('GARD.csv')\n",
    "GARD_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "54ae3242",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GARD.csv d.synonyms has oddly saved string data that cannot be converted directly into a list, this converts that\n",
    "def str2list(string):\n",
    "    string = str(string).replace('[','')\n",
    "    string = string.replace(']','')\n",
    "    string = string.strip()\n",
    "    str_list = string.split(',')\n",
    "    for s in str_list:\n",
    "        s = s.strip()\n",
    "        if s=='nan':\n",
    "            str_list.remove('nan')\n",
    "    return str_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8549d22b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Convert d.synonym strings into lists\n",
    "i=0\n",
    "for i in range(len(GARD_df['d.synonyms'])):\n",
    "    GARD_df['d.synonyms'][i] = str2list(GARD_df['d.synonyms'][i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9680d2f9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d.gard_id</th>\n",
       "      <th>d.name</th>\n",
       "      <th>d.synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6056</th>\n",
       "      <td>GARD:0013731</td>\n",
       "      <td>T-cell prolymphocytic leukemia</td>\n",
       "      <td>[T Cell Prolymphocytic Leukemia]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6057</th>\n",
       "      <td>GARD:0013735</td>\n",
       "      <td>Spastic paraplegia 47</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6058</th>\n",
       "      <td>GARD:0013737</td>\n",
       "      <td>AP-4-Associated Hereditary Spastic Paraplegia</td>\n",
       "      <td>[Severe intellectual disability and progressiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6059</th>\n",
       "      <td>GARD:0013743</td>\n",
       "      <td>Multicentric osteolysis, nodulosis and arthrop...</td>\n",
       "      <td>[Torg-Winchester Syndrome, Torg Syndrome, Nodu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6060</th>\n",
       "      <td>GARD:0013818</td>\n",
       "      <td>Sphingosine phosphate lyase insufficiency synd...</td>\n",
       "      <td>[SPL insufficiency syndrome, SPLIS, Familial s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         d.gard_id                                             d.name  \\\n",
       "6056  GARD:0013731                     T-cell prolymphocytic leukemia   \n",
       "6057  GARD:0013735                              Spastic paraplegia 47   \n",
       "6058  GARD:0013737      AP-4-Associated Hereditary Spastic Paraplegia   \n",
       "6059  GARD:0013743  Multicentric osteolysis, nodulosis and arthrop...   \n",
       "6060  GARD:0013818  Sphingosine phosphate lyase insufficiency synd...   \n",
       "\n",
       "                                             d.synonyms  \n",
       "6056                   [T Cell Prolymphocytic Leukemia]  \n",
       "6057                                                 []  \n",
       "6058  [Severe intellectual disability and progressiv...  \n",
       "6059  [Torg-Winchester Syndrome, Torg Syndrome, Nodu...  \n",
       "6060  [SPL insufficiency syndrome, SPLIS, Familial s...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GARD_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bef3fbf7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gracile syndrome\n",
      "finnish lactic acidosis with hepatic hemosiderosis\n",
      "hyperphalangy-clinodactyly of index finger with pierre robin syndrome\n",
      "partial deletion of the short arm of chromosome 3\n",
      "with partial agenesis of the corpus callosum and arachnoid cysts\n",
      "macrothrombocytopenia and granulocyte inclusions with or without nephritis or sensorineural hearing loss\n",
      "malignant tumors of the central nervous system associated with familial polyposis of the colon\n",
      "lethal autosomal recessive arthrogryposis multiplex congenita with whistling face and calcifications of the nervous system\n",
      "xy disorder of sex development due to luteinizing hormone resistance or luteinizing hormone beta subunit deficiency\n",
      "severe or complete loss of motor function in the lower extremities and lower portions of the trunk\n",
      "17\n",
      "22714\n",
      "10492\n"
     ]
    }
   ],
   "source": [
    "#Set up a new & easier to use list of diseases\n",
    "rowlist = []\n",
    "i=0\n",
    "for i in range(len(GARD_df)):\n",
    "    columnlist=[]\n",
    "    columnlist.append(GARD_df['d.name'][i])\n",
    "    columnlist+=GARD_df['d.synonyms'][i]\n",
    "    rowlist.append(columnlist)\n",
    "\n",
    "#keys are going to be disease names, values are going to be the GARD ID, set up this way bc dictionaries are faster lookup than lists\n",
    "GARD_dict = {}\n",
    "GARD_firstwd_dict = {}\n",
    "\n",
    "#Find out what the length of the longest disease name sequence is, of all names and synonyms\n",
    "max_length = -1\n",
    "for i in range(len(rowlist)):\n",
    "    #Compare length of primary disease name\n",
    "    #dz = str(rowlist[i][0]).strip()\n",
    "    #l_dz = len(dz.split())\n",
    "    #if l_dz>max_length:\n",
    "    #    max_length = l_dz\n",
    "    for j in range(len(rowlist[i])):\n",
    "        if rowlist[i][j] not in GARD_dict.keys():\n",
    "            s = str(rowlist[i][j]).lower().strip()\n",
    "            if len(s.split())>0 and s not in STOPWORDS:\n",
    "                GARD_dict[s] = GARD_df['d.gard_id'][i]\n",
    "                #GARD_firstwd_dict[s.split()[0]] = GARD_df['d.gard_id'][i]\n",
    "                #This will increase the false negative rate a little bit, but decrease the false positive rate tremendously\n",
    "                if s.split()[0] not in STOPWORDS:\n",
    "                    GARD_firstwd_dict[s.split()[0]] = GARD_df['d.gard_id'][i]\n",
    "            #compare length\n",
    "            l = len(s.split())\n",
    "            if l>max_length:\n",
    "                print(s)\n",
    "                max_length = l\n",
    "print(max_length)\n",
    "print(len(GARD_dict))\n",
    "print(len(GARD_firstwd_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bfde0f85",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GARD:0012486'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#This means that the GARD.csv dataset is inaccurate, but I would rather have higher recall than higher precision\n",
    "GARD_dict['corpus callosum']\n",
    "#GARD_firstwd_dict['is']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75ad0e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This will not be run now, will be done after training the model since it is just dictionary lookup\n",
    "def tag_diseases(tokens,labels):   \n",
    "    i=0\n",
    "    while i <len(tokens):\n",
    "        if (len(tokens)-i) < max_length:\n",
    "            compare_length=len(tokens)-i\n",
    "        else:\n",
    "            compare_length = max_length\n",
    "        #Compares longest sequences first and goes down until there is a match\n",
    "        #print('(start compare_length)',compare_length)\n",
    "        exit = False\n",
    "        while compare_length>0:\n",
    "            s = ' '.join(tokens[i:i+compare_length])\n",
    "            for key in GARD_dict.keys():\n",
    "                if key==s.lower():\n",
    "                    labels[i] = 'B-DIS'\n",
    "                    #print(s)\n",
    "                    for j in range(i+1,i+compare_length):\n",
    "                        labels[j] = 'I-DIS'\n",
    "                    #Need to skip over the next few indexes\n",
    "                    #print('(compare_length):',compare_length)\n",
    "                    i+=compare_length-1\n",
    "                    exit =True #this allows you to break out of two loops\n",
    "            #break out of loop in case there are multiple rare diseases in the same sentence\n",
    "            if exit:\n",
    "                break\n",
    "            else:\n",
    "                compare_length-=1\n",
    "        i+=1  \n",
    "    return tokens,labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5fec12a",
   "metadata": {},
   "source": [
    "Testing Cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "29600250",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "with B-DIS\n",
      "partial I-DIS\n",
      "agenesis I-DIS\n",
      "of I-DIS\n",
      "the I-DIS\n",
      "corpus I-DIS\n",
      "callosum I-DIS\n",
      "and I-DIS\n",
      "arachnoid I-DIS\n",
      "cysts I-DIS\n",
      "severe B-DIS\n",
      "or I-DIS\n",
      "complete I-DIS\n",
      "loss I-DIS\n",
      "of I-DIS\n",
      "motor I-DIS\n",
      "function I-DIS\n",
      "in I-DIS\n",
      "the I-DIS\n",
      "lower I-DIS\n",
      "extremities I-DIS\n",
      "and I-DIS\n",
      "lower I-DIS\n",
      "portions I-DIS\n",
      "of I-DIS\n",
      "the I-DIS\n",
      "trunk I-DIS\n",
      "now O\n",
      "we O\n",
      "are O\n",
      "at B-LOC\n",
      "an O\n",
      "impasse O\n",
      "finnish B-DIS\n",
      "lactic I-DIS\n",
      "acidosis I-DIS\n",
      "with I-DIS\n",
      "hepatic I-DIS\n",
      "hemosiderosis I-DIS\n"
     ]
    }
   ],
   "source": [
    "s  = ['with','partial','agenesis','of','the','corpus','callosum','and','arachnoid','cysts','severe','or','complete','loss','of','motor','function','in','the','lower','extremities','and','lower','portions','of','the','trunk','now','we','are','at','an','impasse','finnish','lactic','acidosis','with','hepatic','hemosiderosis']\n",
    "t = ['B-LOC' if i%5==0 else 'O' for i in range(len(s))]\n",
    "\n",
    "#for i in range(len(s)):\n",
    "#    print(s[i],t[i])\n",
    "#print('')\n",
    "s,t = tag_diseases(s,t)\n",
    "print('')\n",
    "for i in range(len(s)):\n",
    "    print(s[i],t[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ae700d30",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "here O\n",
      "is O\n",
      "GRACILE O\n",
      "Syndrome O\n",
      "Right O\n",
      "Here O\n",
      "\n",
      "\n",
      "here O\n",
      "is O\n",
      "GRACILE B-DIS\n",
      "Syndrome I-DIS\n",
      "Right O\n",
      "Here O\n"
     ]
    }
   ],
   "source": [
    "ts = ['here','is','GRACILE','Syndrome','Right','Here']\n",
    "ls = ['O','O','O','O','O','O']\n",
    "for i in range(len(ts)):\n",
    "    print(ts[i],ls[i])\n",
    "print('')\n",
    "ts,ls = tag_diseases(ts,ls)\n",
    "print('')\n",
    "for i in range(len(ts)):\n",
    "    print(ts[i],ls[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7f6a0c",
   "metadata": {},
   "source": [
    "### Tagging Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "296c1607",
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_stats(tokens,labels):\n",
    "    i=1\n",
    "    while i<len(labels)-1:\n",
    "        if 'STAT' in labels[i]:\n",
    "            #Includes <, > number in the statistic\n",
    "            if tokens[i-1]=='<' or tokens[i-1]=='>':\n",
    "                labels[i-1]='B-STAT'\n",
    "                labels[i]='I-STAT'\n",
    "            #Includes greater than, less than, more than, etc. \n",
    "            if tokens[i-1]=='than':\n",
    "                labels[i-2]='B-STAT'\n",
    "                labels[i-1]='I-STAT'\n",
    "                labels[i]='I-STAT'\n",
    "                \n",
    "        #Combines \"This disease affects 1 in 7500 to 1 in 10,000 people\" into a single statistic phrase instead of 2\n",
    "        if 'STAT' in labels[i-1] and 'STAT' in labels[i+1] and 'STAT' not in labels[i]:\n",
    "            if tokens[i] =='to':\n",
    "                labels[i]='I-STAT'\n",
    "                labels[i+1]='I-STAT'\n",
    "            if tokens[i] =='-':\n",
    "                labels[i]='I-STAT'\n",
    "                labels[i+1]='I-STAT'\n",
    "        \n",
    "        #This gets of the type \"prevalence of 2 to 18 per 100,000\"\n",
    "        if labels[i+1]=='B-STAT':\n",
    "            if tokens[i]=='to' or tokens[i]=='-' or tokens[i-1].isdigit():\n",
    "                labels[i-1]='B-STAT'\n",
    "                labels[i]='I-STAT'\n",
    "                labels[i+1]='I-STAT'\n",
    "        i+=1\n",
    "    return tokens,labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c0522916",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function should take in a sentence and output each word in it with a tentative label\n",
    "def tag_NERs(sentence):\n",
    "    \n",
    "    doc = nlp(sentence)\n",
    "    tokens = [token.text for token in doc]\n",
    "    labels = ['O' for token in doc]\n",
    "    \n",
    "    i = 0\n",
    "    for token in doc:\n",
    "        if len(str(token.text).strip())==0:\n",
    "            tokens.pop(i)\n",
    "            labels.pop(i)\n",
    "            \n",
    "        else:\n",
    "            ## Epidemiologic identifier\n",
    "            if token.text.lower() in {'incidence','prevalence','prevalences','prevalence ','incidences','occurrence','occurrences'}:\n",
    "                labels[i] = 'B-EPI'\n",
    "        \n",
    "            ## Location\n",
    "            if token.ent_type_ in {'GPE','LOC'}:\n",
    "                labels[i] = str(token.ent_iob_+'-LOC')\n",
    "            if token.text in {\"worldwide\"}:\n",
    "                labels[i] = 'B-LOC'\n",
    "        \n",
    "            ## Epidemiologic Rates\n",
    "            #This gets stuff of the form 3.5/100\n",
    "            if token.text[0].isdigit() and '/' in token.text:\n",
    "                labels[i] = 'B-STAT'\n",
    "        \n",
    "            #label all percents except those preceding \"confidence interval (CI)\"\n",
    "            if token.ent_type_ in {'PERCENT'}:# and token.text not in {'95', 'CI'}:\n",
    "                if i<len(doc)-2:\n",
    "                    if doc[i+2].text in {'CI','confidence','interval','confidence interval','(CI)','(CI','CI)'}:\n",
    "                        labels[i] = 'O'\n",
    "                        labels[i+1] = 'O'\n",
    "                        labels[i+2] = 'O'\n",
    "                    elif doc[i+1].text in {'CI','confidence','interval','confidence interval','(CI)','(CI','CI)'}:\n",
    "                        labels[i] = 'O'\n",
    "                        labels[i+1] = 'O'\n",
    "                    else:\n",
    "                        labels[i] = str(token.ent_iob_+'-STAT')\n",
    "                elif i<len(doc)-1:\n",
    "                    if doc[i+1].text in {'CI','confidence','interval','confidence interval','(CI)','(CI','CI)'}:\n",
    "                        labels[i] = 'O'\n",
    "                        labels[i+1] = 'O'\n",
    "                    else:\n",
    "                        labels[i] = str(token.ent_iob_+'-STAT')        \n",
    "                else:\n",
    "                    labels[i] = str(token.ent_iob_+'-STAT')\n",
    "        \n",
    "            #These 3 get stuff of the form \"one in 35000\" or \"one in every 23043\"\n",
    "            if (token.text.lower() in {'one','1'} and i<(len(doc)-3)): \n",
    "                if doc[i+3].is_digit:\n",
    "                    labels[i] = 'B-STAT'\n",
    "                    for j in range(i+1,i+4):\n",
    "                        labels[j] = 'I-STAT'\n",
    "            if (token.text.lower() in {'one','1'} and i<(len(doc)-2)): \n",
    "                if doc[i+2].is_digit:\n",
    "                    labels[i] = 'B-STAT'\n",
    "                    labels[i+1] = 'I-STAT'\n",
    "                    labels[i+2] = 'I-STAT'\n",
    "            if (token.text.lower() in {'one','1'} and i<(len(doc)-1)):\n",
    "                if doc[i+1].is_digit:\n",
    "                    labels[i] = 'B-STAT'\n",
    "                    labels[i+1] = 'I-STAT'\n",
    "        \n",
    "            #These should get the ones of the form: 14.1 deaths per 1,000 LBs\n",
    "            #This is a big decision tree, not sure how to write it in fewer lines of code\n",
    "            #Need to get all permutations of \"a b per c d e\" where (a or b) and (c or d) is number and e is anything, but if e does not exist still need to tag a-d as STAT\n",
    "            if token.text.lower() =='per':\n",
    "                #print(i,len(doc))\n",
    "                if i>1:\n",
    "                    if i<len(doc)-3:\n",
    "                        #Resulted in better testing when not validating that words after 'per' are numbers\n",
    "                        if (doc[i-2].is_digit or doc[i-2].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i-1].is_digit or doc[i-1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+2].is_digit or doc[i+2].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+1].is_digit or doc[i+1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}):\n",
    "                            if tokens[i-2] not in STOPWORDS and tokens[i-2] not in PUNCTUATION:\n",
    "                                labels[i-2] = 'B-STAT'\n",
    "                                #labeling also the token after the number\n",
    "                                for j in range(i-1,i+3):\n",
    "                                    labels[j]='I-STAT'\n",
    "                            else:\n",
    "                                labels[i-1] = 'B-STAT'\n",
    "                                #labeling also the token after the number\n",
    "                                for j in range(i,i+3):\n",
    "                                    labels[j]='I-STAT'\n",
    "                            \n",
    "                    if i<len(doc)-2:\n",
    "                        if (doc[i-2].is_digit or doc[i-2].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i-1].is_digit or doc[i-1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+2].is_digit or doc[i+2].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+1].is_digit or doc[i+1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}):\n",
    "                            if tokens[i-2] not in STOPWORDS and tokens[i-2] not in PUNCTUATION:\n",
    "                                labels[i-2] = 'B-STAT'\n",
    "                                #labeling also the token after the number\n",
    "                                for j in range(i-1,i+2):\n",
    "                                    labels[j]='I-STAT'\n",
    "                            else: \n",
    "                                labels[i-1] = 'B-STAT'\n",
    "                                #labeling also the token after the number\n",
    "                                for j in range(i,i+2):\n",
    "                                    labels[j]='I-STAT'\n",
    "                    #The difference between the above and below is in labeling the token immediately after the number\n",
    "                    if i<len(doc)-1:\n",
    "                        if (doc[i-2].is_digit or doc[i-2].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i-1].is_digit or doc[i-1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+2].is_digit or doc[i+2].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+1].is_digit or doc[i+1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}):\n",
    "                            if tokens[i-2] not in STOPWORDS and tokens[i-2] not in PUNCTUATION:\n",
    "                                labels[i-2] = 'B-STAT'\n",
    "                                #labeling also the token after if it is number\n",
    "                                for j in range(i-1,i+1):\n",
    "                                    labels[j]='I-STAT'\n",
    "                            else: \n",
    "                                labels[i-1] = 'B-STAT'\n",
    "                                #labeling also the token after the number\n",
    "                                for j in range(i,i+1):\n",
    "                                    labels[j]='I-STAT'\n",
    "                elif i>0:\n",
    "                    if i<len(doc)-3:\n",
    "                        if (doc[i-1].is_digit or doc[i-1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+2].is_digit or doc[i+2].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+1].is_digit or doc[i+1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}):\n",
    "                \n",
    "                            labels[i-1] = 'B-STAT'\n",
    "                            #labeling also the token after the number\n",
    "                            for j in range(i,i+3):\n",
    "                                labels[j]='I-STAT'\n",
    "                            \n",
    "                    if i<len(doc)-2:\n",
    "                        if (doc[i-1].is_digit or doc[i-1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+2].is_digit or doc[i+2].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+1].is_digit or doc[i+1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}):\n",
    "                        \n",
    "                            labels[i-1] = 'B-STAT'\n",
    "                            #labeling also the token after the number\n",
    "                            for j in range(i,i+2):\n",
    "                                labels[j]='I-STAT'\n",
    "                            \n",
    "                    if i<len(doc)-1:\n",
    "                        if (doc[i-1].is_digit or doc[i-1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}) or (\n",
    "                            doc[i+1].is_digit or doc[i+1].ent_type_ in {'CARDINAL','ORDINAL','QUANTITY','MONEY'}):\n",
    "                    \n",
    "                            labels[i-1] = 'B-STAT'\n",
    "                            #labeling just the number if there is nothing after. \n",
    "                            for j in range(i,i+1):\n",
    "                                labels[j]='I-STAT'\n",
    "            i+=1\n",
    "\n",
    "    if len(tokens) != len(labels):\n",
    "        raise ValueError('Token/Label Length Mismatch')\n",
    "        \n",
    "    if len(tokens)>2 and len(labels)>2:\n",
    "        tokens, labels = combine_stats(tokens,labels)\n",
    "\n",
    "    return tokens, labels #This returns as type Spacy.tokens, need to convert to strings at writing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52ae06d0",
   "metadata": {},
   "source": [
    "### tag_NERs Function Testing Cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b58e5dfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_dictionary = {'s2':'5% of guards run to the watch tower in Budapest where one 6 14.1 deaths per 1,000 LBs here here the king (95% CI 148.3-254.2)',\n",
    "    's3':'Organ damage in sickle cell disease (SCD) is a crucial determinant for disease severity and prognosis. In a previous study, we analyzed the prevalence of SCD-related organ damage and complications in adult sickle cell patients. We now describe a seven-year follow-up of this cohort.All patients from the primary analysis in 2006 (n = 104), were included for follow-up. Patients were screened for SCD-related organ damage and complications (microalbuminuria, renal failure, elevated tricuspid regurgitation flow velocity (TRV) (≥2.5 m/seconds), retinopathy, iron overload, cholelithiasis, avascular osteonecrosis, leg ulcers, acute chest syndrome (ACS), stroke, priapism and admissions for vaso-occlusive crises (VOC) biannually. Upon 7 years of follow-up, progression in the prevalence of avascular osteonecrosis (from 12.5% to 20.4%), renal failure (from 6.7% to 23.4%), retinopathy (from 39.7% to 53.8%) was observed in the whole group. In HbSS/HbSβ0 -thal patients also progression in microalbuminuria (from 34% to 45%) and elevated TRV (from 40% to 48%) was observed while hardly any progression in the prevalence of cholelithiasis, priapism, stroke or chronic ulcers was seen. The proportion of patients with at least one episode of ACS increased in the group of HbSS/HbSβ0 -thal patients from 32% to 49.1%. In conclusion, 62% of the sickle cell patients in this prospective cohort study developed a new SCD-related complication in a comprehensive care setting within 7 years of follow-up. Although the hospital admission rate for VOC remained stable, multiple forms of organ damage increased substantially. These observations underline the need for continued screening for organ damage in all adult patients with SCD.',\n",
    "    's4':\"Uveal melanoma (UM) represents the most prominent primary eye cancer in adults. With an incidence of approximately 5 cases per million individuals annually in the United States, UM could be considered a relatively rare cancer. The 90‑95% of UM cases arise from the choroid. Diagnosis is based mainly on a clinical examination and ancillary tests, with ocular ultrasonography being of greatest value. Differential diagnosis can prove challenging in the case of indeterminate choroidal lesions and, sometimes, monitoring for documented growth may be the proper approach. Fine needle aspiration biopsy tends to be performed with a prognostic purpose, often in combination with radiotherapy. Gene expression profiling has allowed for the grading of UMs into two classes, which feature different metastatic risks. Patients with UM require a specialized multidisciplinary management. Primary tumor treatment can be either enucleation or globe preserving. Usually, enucleation is reserved for larger tumors, while radiotherapy is preferred for small/medium melanomas. The prognosis is unfavorable due to the high mortality rate and high tendency to metastasize. Following the development of metastatic disease, the mortality rate increases to 80% within one year, due to both the absence of an effective treatment and the aggressiveness of the condition. Novel molecular studies have allowed for a better understanding of the genetic and epigenetic mechanisms involved in UM biological activity, which differs compared to skin melanomas. The most commonly mutated genes are GNAQ, GNA11 and BAP1. Research in this field could help to identify effective diagnostic and prognostic biomarkers, as well as novel therapeutic targets.\",\n",
    "    's5':\"This disease affects <1/500,000 baby boys in Russia. It is linked to the   original hemopheliac prince and is a old disease.\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6db6ecd8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "----------------- s2 -------------------------\n",
      "\n",
      "5 B-STAT\n",
      "% I-STAT\n",
      "of O\n",
      "guards O\n",
      "run O\n",
      "to O\n",
      "the O\n",
      "watch O\n",
      "tower O\n",
      "in O\n",
      "Budapest B-LOC\n",
      "where O\n",
      "one B-STAT\n",
      "6 I-STAT\n",
      "14.1 B-STAT\n",
      "deaths I-STAT\n",
      "per I-STAT\n",
      "1,000 I-STAT\n",
      "LBs I-STAT\n",
      "here O\n",
      "here O\n",
      "the O\n",
      "king O\n",
      "( O\n",
      "95 O\n",
      "% O\n",
      "CI O\n",
      "148.3 O\n",
      "- O\n",
      "254.2 O\n",
      ") O\n",
      "\n",
      "----------------- s3 -------------------------\n",
      "\n",
      "Organ O\n",
      "damage O\n",
      "in O\n",
      "sickle O\n",
      "cell O\n",
      "disease O\n",
      "( O\n",
      "SCD O\n",
      ") O\n",
      "is O\n",
      "a O\n",
      "crucial O\n",
      "determinant O\n",
      "for O\n",
      "disease O\n",
      "severity O\n",
      "and O\n",
      "prognosis O\n",
      ". O\n",
      "In O\n",
      "a O\n",
      "previous O\n",
      "study O\n",
      ", O\n",
      "we O\n",
      "analyzed O\n",
      "the O\n",
      "prevalence B-EPI\n",
      "of O\n",
      "SCD O\n",
      "- O\n",
      "related O\n",
      "organ O\n",
      "damage O\n",
      "and O\n",
      "complications O\n",
      "in O\n",
      "adult O\n",
      "sickle O\n",
      "cell O\n",
      "patients O\n",
      ". O\n",
      "We O\n",
      "now O\n",
      "describe O\n",
      "a O\n",
      "seven O\n",
      "- O\n",
      "year O\n",
      "follow O\n",
      "- O\n",
      "up O\n",
      "of O\n",
      "this O\n",
      "cohort O\n",
      ". O\n",
      "All O\n",
      "patients O\n",
      "from O\n",
      "the O\n",
      "primary O\n",
      "analysis O\n",
      "in O\n",
      "2006 O\n",
      "( O\n",
      "n O\n",
      "= O\n",
      "104 O\n",
      ") O\n",
      ", O\n",
      "were O\n",
      "included O\n",
      "for O\n",
      "follow O\n",
      "- O\n",
      "up O\n",
      ". O\n",
      "Patients O\n",
      "were O\n",
      "screened O\n",
      "for O\n",
      "SCD O\n",
      "- O\n",
      "related O\n",
      "organ O\n",
      "damage O\n",
      "and O\n",
      "complications O\n",
      "( O\n",
      "microalbuminuria O\n",
      ", O\n",
      "renal O\n",
      "failure O\n",
      ", O\n",
      "elevated O\n",
      "tricuspid O\n",
      "regurgitation O\n",
      "flow O\n",
      "velocity O\n",
      "( O\n",
      "TRV O\n",
      ") O\n",
      "( O\n",
      "≥2.5 O\n",
      "m O\n",
      "/ O\n",
      "seconds O\n",
      ") O\n",
      ", O\n",
      "retinopathy O\n",
      ", O\n",
      "iron O\n",
      "overload O\n",
      ", O\n",
      "cholelithiasis O\n",
      ", O\n",
      "avascular O\n",
      "osteonecrosis O\n",
      ", O\n",
      "leg O\n",
      "ulcers O\n",
      ", O\n",
      "acute O\n",
      "chest O\n",
      "syndrome O\n",
      "( O\n",
      "ACS O\n",
      ") O\n",
      ", O\n",
      "stroke O\n",
      ", O\n",
      "priapism O\n",
      "and O\n",
      "admissions O\n",
      "for O\n",
      "vaso O\n",
      "- O\n",
      "occlusive O\n",
      "crises O\n",
      "( O\n",
      "VOC O\n",
      ") O\n",
      "biannually O\n",
      ". O\n",
      "Upon O\n",
      "7 O\n",
      "years O\n",
      "of O\n",
      "follow O\n",
      "- O\n",
      "up O\n",
      ", O\n",
      "progression O\n",
      "in O\n",
      "the O\n",
      "prevalence B-EPI\n",
      "of O\n",
      "avascular O\n",
      "osteonecrosis O\n",
      "( O\n",
      "from O\n",
      "12.5 B-STAT\n",
      "% I-STAT\n",
      "to I-STAT\n",
      "20.4 I-STAT\n",
      "% I-STAT\n",
      ") O\n",
      ", O\n",
      "renal O\n",
      "failure O\n",
      "( O\n",
      "from O\n",
      "6.7 B-STAT\n",
      "% I-STAT\n",
      "to I-STAT\n",
      "23.4 I-STAT\n",
      "% I-STAT\n",
      ") O\n",
      ", O\n",
      "retinopathy O\n",
      "( O\n",
      "from O\n",
      "39.7 B-STAT\n",
      "% I-STAT\n",
      "to I-STAT\n",
      "53.8 I-STAT\n",
      "% I-STAT\n",
      ") O\n",
      "was O\n",
      "observed O\n",
      "in O\n",
      "the O\n",
      "whole O\n",
      "group O\n",
      ". O\n",
      "In O\n",
      "HbSS B-LOC\n",
      "/ O\n",
      "HbSβ0 O\n",
      "-thal O\n",
      "patients O\n",
      "also O\n",
      "progression O\n",
      "in O\n",
      "microalbuminuria O\n",
      "( O\n",
      "from O\n",
      "34 B-STAT\n",
      "% I-STAT\n",
      "to I-STAT\n",
      "45 I-STAT\n",
      "% I-STAT\n",
      ") O\n",
      "and O\n",
      "elevated O\n",
      "TRV O\n",
      "( O\n",
      "from O\n",
      "40 B-STAT\n",
      "% I-STAT\n",
      "to I-STAT\n",
      "48 I-STAT\n",
      "% I-STAT\n",
      ") O\n",
      "was O\n",
      "observed O\n",
      "while O\n",
      "hardly O\n",
      "any O\n",
      "progression O\n",
      "in O\n",
      "the O\n",
      "prevalence B-EPI\n",
      "of O\n",
      "cholelithiasis O\n",
      ", O\n",
      "priapism O\n",
      ", O\n",
      "stroke O\n",
      "or O\n",
      "chronic O\n",
      "ulcers O\n",
      "was O\n",
      "seen O\n",
      ". O\n",
      "The O\n",
      "proportion O\n",
      "of O\n",
      "patients O\n",
      "with O\n",
      "at O\n",
      "least O\n",
      "one O\n",
      "episode O\n",
      "of O\n",
      "ACS O\n",
      "increased O\n",
      "in O\n",
      "the O\n",
      "group O\n",
      "of O\n",
      "HbSS B-LOC\n",
      "/ O\n",
      "HbSβ0 O\n",
      "-thal O\n",
      "patients O\n",
      "from O\n",
      "32 B-STAT\n",
      "% I-STAT\n",
      "to I-STAT\n",
      "49.1 I-STAT\n",
      "% I-STAT\n",
      ". O\n",
      "In O\n",
      "conclusion O\n",
      ", O\n",
      "62 B-STAT\n",
      "% I-STAT\n",
      "of O\n",
      "the O\n",
      "sickle O\n",
      "cell O\n",
      "patients O\n",
      "in O\n",
      "this O\n",
      "prospective O\n",
      "cohort O\n",
      "study O\n",
      "developed O\n",
      "a O\n",
      "new O\n",
      "SCD O\n",
      "- O\n",
      "related O\n",
      "complication O\n",
      "in O\n",
      "a O\n",
      "comprehensive O\n",
      "care O\n",
      "setting O\n",
      "within O\n",
      "7 O\n",
      "years O\n",
      "of O\n",
      "follow O\n",
      "- O\n",
      "up O\n",
      ". O\n",
      "Although O\n",
      "the O\n",
      "hospital O\n",
      "admission O\n",
      "rate O\n",
      "for O\n",
      "VOC O\n",
      "remained O\n",
      "stable O\n",
      ", O\n",
      "multiple O\n",
      "forms O\n",
      "of O\n",
      "organ O\n",
      "damage O\n",
      "increased O\n",
      "substantially O\n",
      ". O\n",
      "These O\n",
      "observations O\n",
      "underline O\n",
      "the O\n",
      "need O\n",
      "for O\n",
      "continued O\n",
      "screening O\n",
      "for O\n",
      "organ O\n",
      "damage O\n",
      "in O\n",
      "all O\n",
      "adult O\n",
      "patients O\n",
      "with O\n",
      "SCD O\n",
      ". O\n",
      "\n",
      "----------------- s4 -------------------------\n",
      "\n",
      "Uveal O\n",
      "melanoma O\n",
      "( O\n",
      "UM O\n",
      ") O\n",
      "represents O\n",
      "the O\n",
      "most O\n",
      "prominent O\n",
      "primary O\n",
      "eye O\n",
      "cancer O\n",
      "in O\n",
      "adults O\n",
      ". O\n",
      "With O\n",
      "an O\n",
      "incidence B-EPI\n",
      "of O\n",
      "approximately O\n",
      "5 B-STAT\n",
      "cases I-STAT\n",
      "per I-STAT\n",
      "million I-STAT\n",
      "individuals I-STAT\n",
      "annually O\n",
      "in O\n",
      "the B-LOC\n",
      "United I-LOC\n",
      "States I-LOC\n",
      ", O\n",
      "UM O\n",
      "could O\n",
      "be O\n",
      "considered O\n",
      "a O\n",
      "relatively O\n",
      "rare O\n",
      "cancer O\n",
      ". O\n",
      "The O\n",
      "90‑95 B-STAT\n",
      "% I-STAT\n",
      "of O\n",
      "UM O\n",
      "cases O\n",
      "arise O\n",
      "from O\n",
      "the O\n",
      "choroid O\n",
      ". O\n",
      "Diagnosis O\n",
      "is O\n",
      "based O\n",
      "mainly O\n",
      "on O\n",
      "a O\n",
      "clinical O\n",
      "examination O\n",
      "and O\n",
      "ancillary O\n",
      "tests O\n",
      ", O\n",
      "with O\n",
      "ocular O\n",
      "ultrasonography O\n",
      "being O\n",
      "of O\n",
      "greatest O\n",
      "value O\n",
      ". O\n",
      "Differential O\n",
      "diagnosis O\n",
      "can O\n",
      "prove O\n",
      "challenging O\n",
      "in O\n",
      "the O\n",
      "case O\n",
      "of O\n",
      "indeterminate O\n",
      "choroidal O\n",
      "lesions O\n",
      "and O\n",
      ", O\n",
      "sometimes O\n",
      ", O\n",
      "monitoring O\n",
      "for O\n",
      "documented O\n",
      "growth O\n",
      "may O\n",
      "be O\n",
      "the O\n",
      "proper O\n",
      "approach O\n",
      ". O\n",
      "Fine O\n",
      "needle O\n",
      "aspiration O\n",
      "biopsy O\n",
      "tends O\n",
      "to O\n",
      "be O\n",
      "performed O\n",
      "with O\n",
      "a O\n",
      "prognostic O\n",
      "purpose O\n",
      ", O\n",
      "often O\n",
      "in O\n",
      "combination O\n",
      "with O\n",
      "radiotherapy O\n",
      ". O\n",
      "Gene O\n",
      "expression O\n",
      "profiling O\n",
      "has O\n",
      "allowed O\n",
      "for O\n",
      "the O\n",
      "grading O\n",
      "of O\n",
      "UMs O\n",
      "into O\n",
      "two O\n",
      "classes O\n",
      ", O\n",
      "which O\n",
      "feature O\n",
      "different O\n",
      "metastatic O\n",
      "risks O\n",
      ". O\n",
      "Patients O\n",
      "with O\n",
      "UM O\n",
      "require O\n",
      "a O\n",
      "specialized O\n",
      "multidisciplinary O\n",
      "management O\n",
      ". O\n",
      "Primary O\n",
      "tumor O\n",
      "treatment O\n",
      "can O\n",
      "be O\n",
      "either O\n",
      "enucleation O\n",
      "or O\n",
      "globe O\n",
      "preserving O\n",
      ". O\n",
      "Usually O\n",
      ", O\n",
      "enucleation O\n",
      "is O\n",
      "reserved O\n",
      "for O\n",
      "larger O\n",
      "tumors O\n",
      ", O\n",
      "while O\n",
      "radiotherapy O\n",
      "is O\n",
      "preferred O\n",
      "for O\n",
      "small O\n",
      "/ O\n",
      "medium O\n",
      "melanomas O\n",
      ". O\n",
      "The O\n",
      "prognosis O\n",
      "is O\n",
      "unfavorable O\n",
      "due O\n",
      "to O\n",
      "the O\n",
      "high O\n",
      "mortality O\n",
      "rate O\n",
      "and O\n",
      "high O\n",
      "tendency O\n",
      "to O\n",
      "metastasize O\n",
      ". O\n",
      "Following O\n",
      "the O\n",
      "development O\n",
      "of O\n",
      "metastatic O\n",
      "disease O\n",
      ", O\n",
      "the O\n",
      "mortality O\n",
      "rate O\n",
      "increases B-STAT\n",
      "to I-STAT\n",
      "80 I-STAT\n",
      "% I-STAT\n",
      "within O\n",
      "one O\n",
      "year O\n",
      ", O\n",
      "due O\n",
      "to O\n",
      "both O\n",
      "the O\n",
      "absence O\n",
      "of O\n",
      "an O\n",
      "effective O\n",
      "treatment O\n",
      "and O\n",
      "the O\n",
      "aggressiveness O\n",
      "of O\n",
      "the O\n",
      "condition O\n",
      ". O\n",
      "Novel O\n",
      "molecular O\n",
      "studies O\n",
      "have O\n",
      "allowed O\n",
      "for O\n",
      "a O\n",
      "better O\n",
      "understanding O\n",
      "of O\n",
      "the O\n",
      "genetic O\n",
      "and O\n",
      "epigenetic O\n",
      "mechanisms O\n",
      "involved O\n",
      "in O\n",
      "UM O\n",
      "biological O\n",
      "activity O\n",
      ", O\n",
      "which O\n",
      "differs O\n",
      "compared O\n",
      "to O\n",
      "skin O\n",
      "melanomas O\n",
      ". O\n",
      "The O\n",
      "most O\n",
      "commonly O\n",
      "mutated O\n",
      "genes O\n",
      "are O\n",
      "GNAQ O\n",
      ", O\n",
      "GNA11 O\n",
      "and O\n",
      "BAP1 O\n",
      ". O\n",
      "Research O\n",
      "in O\n",
      "this O\n",
      "field O\n",
      "could O\n",
      "help O\n",
      "to O\n",
      "identify O\n",
      "effective O\n",
      "diagnostic O\n",
      "and O\n",
      "prognostic O\n",
      "biomarkers O\n",
      ", O\n",
      "as O\n",
      "well O\n",
      "as O\n",
      "novel O\n",
      "therapeutic O\n",
      "targets O\n",
      ". O\n",
      "\n",
      "----------------- s5 -------------------------\n",
      "\n",
      "This O\n",
      "disease O\n",
      "affects O\n",
      "< B-STAT\n",
      "1/500,000 I-STAT\n",
      "baby O\n",
      "boys O\n",
      "in O\n",
      "Russia B-LOC\n",
      ". O\n",
      "It O\n",
      "is O\n",
      "linked O\n",
      "to O\n",
      "the O\n",
      "original O\n",
      "hemopheliac O\n",
      "prince O\n",
      "and O\n",
      "is O\n",
      "a O\n",
      "old O\n",
      "disease O\n",
      ". O\n"
     ]
    }
   ],
   "source": [
    "for k, v in test_dictionary.items():\n",
    "    #Test the function\n",
    "    a,b = tag_NERs(v)\n",
    "    print('\\n-----------------',k,'-------------------------\\n')\n",
    "    for i in range(len(a)):\n",
    "        print(a[i], b[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3d858519",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This O\n",
      "disease O\n",
      "affects O\n",
      "< B-STAT\n",
      "1/500,000 I-STAT\n",
      "baby O\n",
      "boys O\n",
      "in O\n",
      "Russia B-LOC\n",
      ". O\n",
      "It O\n",
      "is O\n",
      "linked O\n",
      "to O\n",
      "the O\n",
      "original O\n",
      "hemopheliac O\n",
      "prince O\n",
      "and O\n",
      "is O\n",
      "a O\n",
      "old O\n",
      "disease O\n",
      ". O\n"
     ]
    }
   ],
   "source": [
    "a,b = tag_NERs(remove_html(test_dictionary['s5']))\n",
    "for i in range(len(a)):\n",
    "    print(a[i], b[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ba01612c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This O\n",
      "disease O\n",
      "affects O\n",
      "< B-STAT\n",
      "1/500,000 I-STAT\n",
      "baby O\n",
      "boys O\n",
      "in O\n",
      "Russia B-LOC\n",
      ". O\n",
      "\n",
      "It O\n",
      "is O\n",
      "linked O\n",
      "to O\n",
      "the O\n",
      "original O\n",
      "hemopheliac O\n",
      "prince O\n",
      "and O\n",
      "is O\n",
      "a O\n",
      "old O\n",
      "disease O\n",
      ". O\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sentences = tokenize.sent_tokenize(remove_html(test_dictionary['s5']))\n",
    "\n",
    "for sent in sentences:\n",
    "    a,b = tag_NERs(sent)\n",
    "    for i in range(len(a)):\n",
    "        print(a[i], b[i])\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d1dc9e76",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "abstracts = abstracts_df['abstract'].to_list()\n",
    "for i in range(len(abstracts)):\n",
    "    abstracts[i] = tokenize.sent_tokenize(remove_html(abstracts[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "14ed0a86",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Background O\n",
      "Kindler O\n",
      "poikiloderma B-DIS\n",
      "is O\n",
      "an O\n",
      "inherited B-DIS\n",
      "autosomal O\n",
      "genodermatosis O\n",
      "characterized O\n",
      "by O\n",
      "blistering O\n",
      "of O\n",
      "the O\n",
      "epidermis O\n",
      "and O\n",
      "mucosae O\n",
      ". O\n",
      "\n",
      "Its O\n",
      "prevalence B-EPI\n",
      "is O\n",
      "unknown O\n",
      ". O\n",
      "\n",
      "Case O\n",
      "report O\n",
      "We O\n",
      "monitored O\n",
      "two O\n",
      "brothers O\n",
      "suffering O\n",
      "from O\n",
      "this O\n",
      "pathology O\n",
      ". O\n",
      "\n",
      "Oral O\n",
      "manifestations O\n",
      "mainly O\n",
      "take O\n",
      "the O\n",
      "form O\n",
      "of O\n",
      "periodontal O\n",
      "lesions O\n",
      ". O\n",
      "\n",
      "In O\n",
      "our O\n",
      "patients O\n",
      "we O\n",
      "noted O\n",
      "gingivitis O\n",
      "progressing O\n",
      "to O\n",
      "periodontitis O\n",
      "as O\n",
      "follow O\n",
      "- O\n",
      "up O\n",
      "care O\n",
      "was O\n",
      "not O\n",
      "effective O\n",
      ". O\n",
      "\n",
      "We O\n",
      "also O\n",
      "diagnosed O\n",
      "enamel B-DIS\n",
      "hypoplasia I-DIS\n",
      ", O\n",
      "described O\n",
      "more O\n",
      "rarely O\n",
      "in O\n",
      "this O\n",
      "pathology O\n",
      ". O\n",
      "\n",
      "Conclusion O\n",
      "Periodontitis O\n",
      "in O\n",
      "Kindler B-DIS\n",
      "Syndrome I-DIS\n",
      "responds O\n",
      "to O\n",
      "maintenance O\n",
      "therapy O\n",
      ", O\n",
      "but O\n",
      "the O\n",
      "absence B-DIS\n",
      "of I-DIS\n",
      "surveillance O\n",
      "is O\n",
      "penalized O\n",
      "by O\n",
      "a O\n",
      "deterioration O\n",
      "in O\n",
      "periodontal O\n",
      "condition O\n",
      "and O\n",
      "complication O\n",
      "of O\n",
      "management O\n",
      ". O\n",
      "\n",
      "All O\n",
      "restorative O\n",
      ", O\n",
      "endodontic O\n",
      ", O\n",
      "surgical O\n",
      ", O\n",
      "periodontal O\n",
      "and O\n",
      "orthodontic O\n",
      "treatments O\n",
      "should O\n",
      "be O\n",
      "performed O\n",
      "with O\n",
      "appropriate O\n",
      "precautions O\n",
      ". O\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Test this again and again with random sampling\n",
    "import random\n",
    "for sentence in random.choice(abstracts):\n",
    "    a,b = tag_NERs(sentence)\n",
    "    a,b = tag_diseases(a,b)\n",
    "    for i in range(len(a)):\n",
    "        print(a[i], b[i])\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80ddaa10",
   "metadata": {},
   "source": [
    "## (4) Split the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6d52aea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "abstract_sents = [tokenize.sent_tokenize(abstract) for abstract in [remove_html(abstract) for abstract in abstracts_df['abstract']]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "81e8c3ec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Background The incidence of hydrocephalus in the spinal muscular atrophy (SMA) population relative to the general population is currently unknown.', 'Since the approval of nusinersen, an intrathecally administered drug for SMA, a small number of hydrocephalus cases among nusinersen users have been reported.', \"Currently, the incidence of hydrocephalus in untreated SMA patients is not available, thereby making it difficult to determine if hydrocephalus is a side effect of nusinersen or part of SMA's natural history.\", 'This retrospective, matched cohort study used electronic health records (EHRs) to estimate and compare the incidence of hydrocephalus in both SMA patients and matched non-SMA controls in the time period prior to the approval of nusinersen.', 'Methods The U.S. Optum® de-identified EHR database contains records for approximately 100 million persons.', 'The current study period spanned January 1, 2007-December 22, 2016.', 'Patients with SMA were identified by one or more International Classification of Diseases (ICD)-9 and/or ICD-10 codes for SMA appearing as primary, admission, or discharge diagnoses, without a pregnancy diagnostic code in the 1-year time before and after the first occurrence of SMA.', 'The first occurrence of SMA defined the index date and non-SMA controls were matched to cases.', 'Incident cases of hydrocephalus were identified with one or more ICD-9 and/or ICD-10 code for any type of hydrocephalus following the index date.', 'Hydrocephalus incidence rates per person-months and the incidence rate ratio comparing SMA cases with non-SMA controls were calculated.', 'Results There were 5354 SMA cases and an equal number of\\xa0matched non-SMA controls.', 'Incident hydrocephalus events were identified in 42 SMA cases and 9 non-SMA controls.', 'Hydrocephalus incidence rates per 100,000 person-months were 15.5 (95% CI: 11.2-20.9) among SMA cases and 3.3 (95% CI: 1.5-6.3) among non-SMA controls.', 'The incidence rate ratio was 4.7 (95% CI: 2.4-10.2).', 'Conclusions Based on this retrospective analysis utilizing US EHR data, SMA patients had an approximately fourfold increased risk of hydrocephalus compared with non-SMA controls in the era preceding nusinersen treatment.', 'This study may assist in properly evaluating adverse events in nusinersen-treated SMA patients.']\n",
      "Background The incidence of hydrocephalus in the spinal muscular atrophy (SMA) population relative to the general population is currently unknown.\n"
     ]
    }
   ],
   "source": [
    "for abstract in abstract_sents:\n",
    "    print(abstract)\n",
    "    for sentence in abstract:\n",
    "        print(sentence)\n",
    "        break\n",
    "    break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c9abd32d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50 570\n",
      "456 114\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "dev_set, test_set = train_test_split(abstract_sents, test_size=50,random_state=5)\n",
    "print(len(test_set),len(dev_set))\n",
    "\n",
    "#dev set is just a holder and will not be saved\n",
    "\n",
    "train_set, val_set = train_test_split(dev_set, train_size=0.8,random_state=6)\n",
    "print(len(train_set),len(val_set))\n",
    "del dev_set"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8954dada",
   "metadata": {},
   "source": [
    "Actually map the tags onto abstract_sents and count how many annotations there are"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f0da807a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step: 250\n",
      "Step: 500\n",
      "22 22\n",
      "number of annotations 13910 13910\n"
     ]
    }
   ],
   "source": [
    "def tag_abstracts(abstract_sents):\n",
    "    all_tokens, all_labels = [],[]\n",
    "    num_tokens,num_labels =0,0\n",
    "    i=0\n",
    "    for abstract in abstract_sents:\n",
    "        abstract_tokens, abstract_labels = [],[]\n",
    "        for sentence in abstract:\n",
    "            sentence_tokens, sentence_labels = tag_NERs(sentence)\n",
    "            num_tokens+=len(sentence_tokens)\n",
    "            num_labels+=len(sentence_labels)\n",
    "            abstract_tokens.append(sentence_tokens)\n",
    "            abstract_labels.append(sentence_labels)\n",
    "            #Count cause it can take a long time\n",
    "            i+=1\n",
    "            if i%250==0:\n",
    "                print('Step:',i)\n",
    "        all_tokens.append(abstract_tokens)\n",
    "        all_labels.append(abstract_labels)\n",
    "            \n",
    "    print(len(sentence_tokens),len(sentence_labels))\n",
    "    print('number of annotations',num_tokens,num_labels)\n",
    "    return all_tokens, all_labels\n",
    "tokens, labels = tag_abstracts(test_set)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4de30f21",
   "metadata": {},
   "source": [
    "## (5) Save into the correct format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "77ec7379",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step: 250\n",
      "Step: 500\n",
      "Step: 750\n",
      "Step: 1000\n",
      "Step: 1250\n",
      "Step: 1500\n",
      "Step: 1750\n",
      "Step: 2000\n",
      "Step: 2250\n",
      "Step: 2500\n",
      "Step: 2750\n",
      "Step: 3000\n",
      "Step: 3250\n",
      "Step: 3500\n",
      "Step: 3750\n",
      "Step: 4000\n",
      "Step: 4250\n",
      "Step: 4500\n",
      "12 12\n",
      "number of annotations 117888 117888\n",
      "abstract num 0 done\n",
      "abstract num 50 done\n",
      "abstract num 100 done\n",
      "abstract num 150 done\n",
      "abstract num 200 done\n",
      "abstract num 250 done\n",
      "abstract num 300 done\n",
      "abstract num 350 done\n",
      "abstract num 400 done\n",
      "abstract num 450 done\n"
     ]
    }
   ],
   "source": [
    "with open('epi_train_setV2.tsv', \"w\") as f:\n",
    "    abstracts, label_list = tag_abstracts(train_set)\n",
    "    for i in range(len(abstracts)): #For abstract in abstracts\n",
    "        for j in range(len(abstracts[i])): #for sentence in abstract\n",
    "            for k in range(len(abstracts[i][j])): #for token in sentence\n",
    "                output = str(abstracts[i][j][k]) +'\\t' +str(label_list[i][j][k])+'\\n'\n",
    "                f.write(output)\n",
    "            f.write('\\n')\n",
    "        #f.write('\\n')\n",
    "        if i%50==0:\n",
    "            print('abstract num',i,'done')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "703518d8",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step: 250\n",
      "Step: 500\n",
      "Step: 750\n",
      "Step: 1000\n",
      "34 34\n",
      "number of annotations 31262 31262\n",
      "abstract num 0 done\n",
      "abstract num 50 done\n",
      "abstract num 100 done\n"
     ]
    }
   ],
   "source": [
    "with open('epi_val_setV2.tsv', \"w\") as f:\n",
    "    abstracts, label_list = tag_abstracts(val_set)\n",
    "    for i in range(len(abstracts)): #For abstract in abstracts\n",
    "        for j in range(len(abstracts[i])): #for sentence in abstract\n",
    "            for k in range(len(abstracts[i][j])): #for token in sentence\n",
    "                output = str(abstracts[i][j][k]) +'\\t' +str(label_list[i][j][k])+'\\n'\n",
    "                f.write(output)\n",
    "            f.write('\\n')\n",
    "        #f.write('\\n')\n",
    "        if i%50==0:\n",
    "            print('abstract num',i,'done')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "997f1610",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step: 250\n",
      "Step: 500\n",
      "22 22\n",
      "number of annotations 13910 13910\n",
      "abstract num 0 done\n",
      "abstract num 10 done\n",
      "abstract num 20 done\n",
      "abstract num 30 done\n",
      "abstract num 40 done\n"
     ]
    }
   ],
   "source": [
    "with open('epi_test_setV2.tsv', \"w\") as f:\n",
    "    abstracts, label_list = tag_abstracts(test_set)\n",
    "    for i in range(len(abstracts)): #For abstract in abstracts\n",
    "        for j in range(len(abstracts[i])): #for sentence in abstract\n",
    "            for k in range(len(abstracts[i][j])): #for token in sentence\n",
    "                output = str(abstracts[i][j][k]) +'\\t' +str(label_list[i][j][k])+'\\n'\n",
    "                f.write(output)\n",
    "            f.write('\\n')\n",
    "        #f.write('\\n')\n",
    "        if i%10==0:\n",
    "            print('abstract num',i,'done')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf05ce1f",
   "metadata": {},
   "source": [
    "OLD function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfac2590",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "#This will not be run now, will be done after training the model since it is just dictionary lookup\n",
    "def tag_diseases(tokens,labels):   \n",
    "    i=0\n",
    "    while i <len(tokens)-1:\n",
    "    #for i in range(len(tokens)):\n",
    "        print(i)\n",
    "        if tokens[i].lower() in GARD_firstwd_dict.keys():\n",
    "            #print(tokens[i])\n",
    "            if (len(tokens)-i) < max_length:\n",
    "                compare_length=len(tokens)-i\n",
    "            else:\n",
    "                compare_length = max_length\n",
    "            #Compares longest sequences first and goes down until there is a match\n",
    "            while compare_length>0:\n",
    "                s = ' '.join(tokens[i:i+compare_length])\n",
    "                for key in GARD_dict.keys():\n",
    "                    if key==s.lower():\n",
    "                        labels[i] = 'B-DIS'\n",
    "                        print(s)\n",
    "                        for j in range(i+1,i+compare_length):\n",
    "                            labels[j] = 'I-DIS'\n",
    "                        #Need to skip over the next few indexes\n",
    "                        print(compare_length)\n",
    "                        i+=compare_length\n",
    "                        #only want to break out of innermost loop in case there are multiple rare diseases in the same sentence\n",
    "                        break\n",
    "                compare_length-=1\n",
    "                \n",
    "                #if s.lower() in GARD_dict.keys():\n",
    "                #    labels[i] = 'B-DIS'\n",
    "                #    print(s)\n",
    "                #    for j in range(i+1,i+compare_length):\n",
    "                #        labels[j] = 'I-DIS'\n",
    "                #    #Need to skip over the next few indexes\n",
    "                #    i+=compare_length\n",
    "                #    #only want to break out of innermost loop in case there are multiple rare diseases in the same sentence\n",
    "                #    break\n",
    "                #else:\n",
    "                #    compare_length-=1\n",
    "                \n",
    "        i+=1  \n",
    "    return tokens,labels\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
